From 5eb334116394d585cf386492a162088e5d86b03c Mon Sep 17 00:00:00 2001
From: Michael W <baddaemon87@gmail.com>
Date: Sun, 28 Jan 2018 19:17:34 +0100
Subject: [PATCH 092/249] wiki: Add validation for peripherals

* Now we have validation we can finally keep the peripherals clean
* Allow a common set of peripherals
* Make the field required
* Fix devices to validate

Change-Id: I2a1ac6c0eb7fe41e85b7d9e07b34a70f0c0ba325
---
 _data/devices/A6020.yml             |  2 +-
 _data/devices/FP2.yml               |  2 +-
 _data/devices/athene.yml            |  2 +-
 _data/devices/chagallwifi.yml       |  4 +-
 _data/devices/cheeseburger.yml      |  3 +-
 _data/devices/clark.yml             |  2 +-
 _data/devices/d855.yml              |  2 +-
 _data/devices/deb.yml               |  2 +-
 _data/devices/ether.yml             |  2 +-
 _data/devices/flo.yml               |  2 +-
 _data/devices/flounder.yml          |  2 +-
 _data/devices/flounder_lte.yml      |  2 +-
 _data/devices/foster.yml            |  2 +-
 _data/devices/fugu.yml              |  2 +-
 _data/devices/ghost.yml             |  2 +-
 _data/devices/gohan.yml             |  2 +-
 _data/devices/h811.yml              |  3 +-
 _data/devices/h815.yml              |  2 +-
 _data/devices/h830.yml              |  2 +-
 _data/devices/h850.yml              |  2 +-
 _data/devices/h870.yml              |  2 +-
 _data/devices/h910.yml              |  3 +-
 _data/devices/h918.yml              |  3 +-
 _data/devices/hammerhead.yml        |  2 +-
 _data/devices/hero2lte.yml          |  3 +-
 _data/devices/herolte.yml           |  4 +-
 _data/devices/hiaeuhl.yml           |  2 +-
 _data/devices/hiaeul.yml            |  2 +-
 _data/devices/himaul.yml            |  2 +-
 _data/devices/himawl.yml            |  2 +-
 _data/devices/hydrogen.yml          |  2 +-
 _data/devices/i9100.yml             |  2 +-
 _data/devices/i9305.yml             |  2 +-
 _data/devices/ido.yml               |  4 +-
 _data/devices/ivy.yml               |  2 +-
 _data/devices/jag3gds.yml           |  2 +-
 _data/devices/jagnm.yml             |  2 +-
 _data/devices/jfltevzw.yml          |  3 +-
 _data/devices/jfltexx.yml           |  3 +-
 _data/devices/k3gxx.yml             |  2 +-
 _data/devices/karin.yml             |  2 +-
 _data/devices/karin_windy.yml       |  2 +-
 _data/devices/kccat6.yml            |  3 +-
 _data/devices/kenzo.yml             |  2 +-
 _data/devices/kingdom.yml           |  2 +-
 _data/devices/kipper.yml            |  2 +-
 _data/devices/kiwi.yml              |  2 +-
 _data/devices/klimtwifi.yml         |  4 +-
 _data/devices/klte.yml              |  3 +-
 _data/devices/kltechn.yml           |  3 +-
 _data/devices/kltechnduo.yml        |  3 +-
 _data/devices/klteduos.yml          |  3 +-
 _data/devices/kltedv.yml            |  3 +-
 _data/devices/kltekdi.yml           |  3 +-
 _data/devices/kltekor.yml           |  3 +-
 _data/devices/ks01lte.yml           |  3 +-
 _data/devices/land.yml              |  2 +-
 _data/devices/lentislte.yml         |  3 +-
 _data/devices/libra.yml             |  2 +-
 _data/devices/lithium.yml           |  2 +-
 _data/devices/ls990.yml             |  2 +-
 _data/devices/ls997.yml             |  2 +-
 _data/devices/lt03lte.yml           |  2 +-
 _data/devices/lux.yml               |  2 +-
 _data/devices/m216.yml              |  2 +-
 _data/devices/m7.yml                |  2 +-
 _data/devices/m7vzw.yml             |  2 +-
 _data/devices/m8.yml                |  2 +-
 _data/devices/m8d.yml               |  2 +-
 _data/devices/manta.yml             |  2 +-
 _data/devices/maserati.yml          |  2 +-
 _data/devices/mido.yml              |  2 +-
 _data/devices/mondrianwifi.yml      |  2 +-
 _data/devices/n3.yml                |  2 +-
 _data/devices/nx512j.yml            |  2 +-
 _data/devices/onyx.yml              |  2 +-
 _data/devices/paella.yml            |  2 +-
 _data/devices/piccolo.yml           |  2 +-
 _data/devices/pme.yml               |  2 +-
 _data/devices/r5.yml                |  2 +-
 _data/devices/r7plus.yml            |  2 +-
 _data/devices/r7sf.yml              |  2 +-
 _data/devices/roth.yml              |  2 +-
 _data/devices/serrano3gxx.yml       |  2 +-
 _data/devices/serranodsdd.yml       |  2 +-
 _data/devices/serranoltexx.yml      |  2 +-
 _data/devices/shamu.yml             |  2 +-
 _data/devices/shieldtablet.yml      |  2 +-
 _data/devices/sumire.yml            |  3 +-
 _data/devices/suzuran.yml           |  2 +-
 _data/devices/t6.yml                |  2 +-
 _data/devices/t6vzw.yml             |  2 +-
 _data/devices/tenshi.yml            |  2 +-
 _data/devices/us996.yml             |  2 +-
 _data/devices/us997.yml             |  2 +-
 _data/devices/v400.yml              |  2 +-
 _data/devices/v410.yml              |  2 +-
 _data/devices/vegetalte.yml         |  2 +-
 _data/devices/vs995.yml             |  2 +-
 _data/devices/w5.yml                |  2 +-
 _data/devices/zerofltexx.yml        |  2 +-
 _includes/deviceinfo.html           | 20 +++++++---
 _includes/snippets/peripherals.html |  5 +++
 device_sample/sample.yml            |  2 +-
 test/schema-06.yml                  | 59 +++++++++++++++++++++++++++++
 105 files changed, 203 insertions(+), 111 deletions(-)
 create mode 100644 _includes/snippets/peripherals.html

diff --git a/_data/devices/A6020.yml b/_data/devices/A6020.yml
index 20b4171..24d43f7 100644
--- a/_data/devices/A6020.yml
+++ b/_data/devices/A6020.yml
@@ -27,7 +27,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '850 900 1900 2100 MHz HSDPA (Europe/Asia), 850 900 1700(AWS) 1900 2100 MHz HSDPA (Latin America)' }
 - {tech: 4G, bands: '1(2100) 3(1800) 5(850) 7(2600) 8(900) 20(800) LTE (Europe/Asia), 2(1900) 3(1800) 4(1700/2100) 7(2600) 28(700) LTE (Latin America)' }
-peripherals: [Dual SIM, Accelerometer, Gyroscope, Proximity sensor, Compass, GPS, FM Radio]
+peripherals: [Dual SIM, Accelerometer, Gyroscope, Proximity sensor, Compass, GPS, FM radio]
 ram: 2/3 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Up</kbd> + <kbd>Volume Down</kbd> and <kbd>Power</kbd>.
 release: 2016-04
diff --git a/_data/devices/FP2.yml b/_data/devices/FP2.yml
index ebf64c2..625bd90 100644
--- a/_data/devices/FP2.yml
+++ b/_data/devices/FP2.yml
@@ -26,7 +26,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900' }
 - {tech: 3G, bands: '8(900) 2(1900) 1(2100) ' }
 - {tech: 4G, bands: '3(1800) 7(2600) 20(800)' }
-peripherals: [3-axis Compass, 3D Accelerometer, 3D Gyroscope, Ambient light, Dual SIM, FM radio, GPS, GLONASS, Proximity sensor]
+peripherals: [Compass, Accelerometer, Gyroscope, Light sensor, Dual SIM, FM radio, GPS, GLONASS, Proximity sensor]
 ram: 2 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Up</kbd> + <kbd>Power</kbd>. Release when boot logo appears.
 release: 2015-12
diff --git a/_data/devices/athene.yml b/_data/devices/athene.yml
index 4aa9bb7..bd2f4e8 100644
--- a/_data/devices/athene.yml
+++ b/_data/devices/athene.yml
@@ -26,7 +26,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM, 1900 MHz CDMA' }
 - {tech: 3G, bands: '850 900 1700 1900 2100 MHz UMTS' }
 - {tech: 4G, bands: '700 850 1700 1800 1900 2600 MHz' }
-peripherals: ['']
+peripherals: None
 ram: 2/3/4 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> + <kbd>Power</kbd>.
   On the next screen use <kbd>Volume Down</kbd> to scroll to recovery and then press <kbd>Volume Up</kbd> to select.
diff --git a/_data/devices/chagallwifi.yml b/_data/devices/chagallwifi.yml
index 96b66fa..3abaf96 100644
--- a/_data/devices/chagallwifi.yml
+++ b/_data/devices/chagallwifi.yml
@@ -26,7 +26,9 @@ network:
 - {tech: 2G, bands: '' }
 - {tech: 3G, bands: '' }
 - {tech: 4G, bands: '' }
-peripherals: [Accelerometer, Compass, Fingerprint reader (not supported), GPS, Gyroscope]
+peripherals:
+- SM-T800: [Accelerometer, Compass, GPS, Gyroscope]
+- Unsupported: [Fingerprint reader]
 ram: 3 GB
 recovery_boot: With the device powered off, hold <kbd>Home</kbd> + <kbd>Volume Up</kbd> + <kbd>Power</kbd>.
 release: 2014
diff --git a/_data/devices/cheeseburger.yml b/_data/devices/cheeseburger.yml
index 2b87b75..5307c82 100644
--- a/_data/devices/cheeseburger.yml
+++ b/_data/devices/cheeseburger.yml
@@ -26,7 +26,8 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM, BC0 CDMA EVDO' }
 - {tech: 3G, bands: '850 900 1700(AWS) 1900 2100 MHz HSDPA' }
 - {tech: 4G, bands: '1 2 4 5 7 8 12 17 18 19 20 25 26 28 29 30 34 38 39 40 41 66' }
-peripherals: [Dual SIM, Fingerprint reader, Accelerometer, Gyroscope, Proximity sensor, Compass, NFC, Hall sensor, Sensor hub, Ambient light sensor]
+peripherals: [Dual SIM, Fingerprint reader, Accelerometer, Gyroscope, Proximity sensor, Compass, NFC,
+  Hall effect sensor, Sensor hub, Light sensor]
 ram: 6/8 GB LPDDR4X
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> + <kbd>Power</kbd>.
 release: 2017-06
diff --git a/_data/devices/clark.yml b/_data/devices/clark.yml
index 01f06d2..3797f46 100644
--- a/_data/devices/clark.yml
+++ b/_data/devices/clark.yml
@@ -24,7 +24,7 @@ network:
 - {tech: 2G, bands: '' }
 - {tech: 3G, bands: '' }
 - {tech: 4G, bands: '' }
-peripherals: ['']
+peripherals: None
 ram: 3 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> + <kbd>Power</kbd>.
   On the next screen use <kbd>Volume Down</kbd> to scroll to recovery and then press <kbd>Volume Up</kbd> to select.
diff --git a/_data/devices/d855.yml b/_data/devices/d855.yml
index f2eb7dc..5dd8d72 100644
--- a/_data/devices/d855.yml
+++ b/_data/devices/d855.yml
@@ -25,7 +25,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '850 900 1900 2100 MHz HSDPA' }
 - {tech: 4G, bands: '700 800 900 1800 2100 2300 2600 MHz LTE' }
-peripherals: [Accelerometer, Compass, Gyroscope, Proximity sensor, FM Radio]
+peripherals: [Accelerometer, Compass, Gyroscope, Proximity sensor, FM radio]
 ram: 2/3 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> + <kbd>Power</kbd>.
 recovery_partition: /dev/block/platform/msm_sdcc.1/by-name/recovery
diff --git a/_data/devices/deb.yml b/_data/devices/deb.yml
index 857360a..2aa4925 100644
--- a/_data/devices/deb.yml
+++ b/_data/devices/deb.yml
@@ -27,7 +27,7 @@ network:
 - {tech: 3G, bands: '850 900 1700 1900 2100 MHz HSDPA' }
 - {tech: 4G, bands: '1(2100) 2(1900) 3(1800) 4(1700/2100) 5(850) 7(2600) 20(800) (EU), 1(2100) 2(1900) 3(1800) 4(1700/2100) 5(850) 13(700) 17(700) (North America)' }
 no_oem_unlock_switch: true
-peripherals: ['']
+peripherals: None
 ram: 2 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> + <kbd>Power</kbd> until the bootloader appears, then release the buttons.
   Use the volume buttons to scroll to "RECOVERY" and press <kbd>Power</kbd> to select.
diff --git a/_data/devices/ether.yml b/_data/devices/ether.yml
index dcc70d4..d43c080 100644
--- a/_data/devices/ether.yml
+++ b/_data/devices/ether.yml
@@ -26,7 +26,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '850 900 1700(AWS) 1800 1900 2100 MHz HSPA+' }
 - {tech: 4G, bands: '1(2100) 2(1900) 3(1800) 4(1700/2100) 5(850) 7(2600) 8(900) 12(700) 17(700) 19(850) 20(800) 28(700) LTE(Cat. 4)' }
-peripherals: [A-GPS, Accelerometer, Compass, GPS, Gyroscope, Light sensor, Magnetic sensor, Proximity sensor, Fingerprint reader, NFC]
+peripherals: [A-GPS, Accelerometer, Compass, GPS, Gyroscope, Light sensor, Hall effect sensor, Proximity sensor, Fingerprint reader, NFC]
 ram: 3 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> + <kbd>Power</kbd>.
 release: 2016-02
diff --git a/_data/devices/flo.yml b/_data/devices/flo.yml
index 5765953..78ba2b0 100644
--- a/_data/devices/flo.yml
+++ b/_data/devices/flo.yml
@@ -24,7 +24,7 @@ network:
 - {tech: 3G, bands: '' }
 - {tech: 4G, bands: '' }
 no_oem_unlock_switch: true
-peripherals: ['']
+peripherals: None
 ram: 2 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> + <kbd>Power</kbd> until the bootloader appears, then release the buttons.
   Use the volume buttons to scroll to "RECOVERY" and press <kbd>Power</kbd> to select.
diff --git a/_data/devices/flounder.yml b/_data/devices/flounder.yml
index b64a1a8..3cc57d9 100644
--- a/_data/devices/flounder.yml
+++ b/_data/devices/flounder.yml
@@ -26,7 +26,7 @@ network:
 - {tech: 2G, bands: '' }
 - {tech: 3G, bands: '' }
 - {tech: 4G, bands: '' }
-peripherals: [Accelerometer, Light sensor, Gyroscope, Hall sensor, GPS]
+peripherals: [Accelerometer, Light sensor, Gyroscope, Hall effect sensor, GPS]
 ram: 2 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> + <kbd>Power</kbd> until the bootloader appears, then release the buttons.
   Use the volume buttons to scroll to "RECOVERY" and press <kbd>Power</kbd> to select.
diff --git a/_data/devices/flounder_lte.yml b/_data/devices/flounder_lte.yml
index 48a27d1..fe070cc 100644
--- a/_data/devices/flounder_lte.yml
+++ b/_data/devices/flounder_lte.yml
@@ -27,7 +27,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '850 900 AWS 1900 2100 MHz HSPA+' }
 - {tech: 4G, bands: 'B1 B2 B3 B4 B5 B7 B8 B20 LTE' }
-peripherals: [Accelerometer, Light sensor, Gyroscope, Hall sensor, GPS]
+peripherals: [Accelerometer, Light sensor, Gyroscope, Hall effect sensor, GPS]
 ram: 2 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> + <kbd>Power</kbd> until the bootloader appears, then release the buttons.
   Use the volume buttons to scroll to "RECOVERY" and press <kbd>Power</kbd> to select.
diff --git a/_data/devices/foster.yml b/_data/devices/foster.yml
index 3727bbd..56deaea 100644
--- a/_data/devices/foster.yml
+++ b/_data/devices/foster.yml
@@ -24,7 +24,7 @@ network:
 - {tech: 3G, bands: '' }
 - {tech: 4G, bands: '' }
 no_oem_unlock_switch: true
-peripherals: ['']
+peripherals: None
 ram: 3 GB
 recovery_boot: With the device powered off, tap <kbd>Power</kbd>, then hold <kbd>Power</kbd> until a menu is displayed.
   Tap <kbd>Power</kbd> until recovery is selected. Press <kbd>Power</kbd> for two seconds to select.
diff --git a/_data/devices/fugu.yml b/_data/devices/fugu.yml
index 5c5a245..5db4a6f 100644
--- a/_data/devices/fugu.yml
+++ b/_data/devices/fugu.yml
@@ -24,7 +24,7 @@ network:
 - {tech: 3G, bands: '' }
 - {tech: 4G, bands: '' }
 no_oem_unlock_switch: true
-peripherals: ['']
+peripherals: None
 ram: 1 GB
 recovery_boot: With the device unplugged, plug in the device while holding the <kbd>Bottom Button</kbd> until a menu is displayed.
   Tap the <kbd>Bottom Button</kbd> until recovery is selected. Press the <kbd>Bottom Button</kbd> for two seconds to select.
diff --git a/_data/devices/ghost.yml b/_data/devices/ghost.yml
index 1c8fc4d..49c5497 100644
--- a/_data/devices/ghost.yml
+++ b/_data/devices/ghost.yml
@@ -26,7 +26,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM (All models), 800 1900 MHz CDMA' }
 - {tech: 3G, bands: '850 900 1800 1900 2100 MHz HSDPA (EU and AT&T), 850 900 1700 1900 2100 MHz HSDPA (T-Mobile)' }
 - {tech: 4G, bands: '2(1900) 4(1700/2100) 5(850) 17(700) LTE (AT&T), 2(1900) 4(1700/2100) 17(700) LTE (T-Mobile), 3(1800) 7(2600) 20(800) LTE (EU)' }
-peripherals: [Motion Coprocessor, Audio Coprocessor, Accelerometer, Gyroscope, Proximity sensor, Compass, Hall sensor, GPS, NFC]
+peripherals: [Motion Coprocessor, Audio Coprocessor, Accelerometer, Gyroscope, Proximity sensor, Compass, Hall effect sensor, GPS, NFC]
 ram: 2 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> + <kbd>Power</kbd> and select recovery.
 release: 2013-08-23
diff --git a/_data/devices/gohan.yml b/_data/devices/gohan.yml
index 6109a20..46b1ed5 100644
--- a/_data/devices/gohan.yml
+++ b/_data/devices/gohan.yml
@@ -25,7 +25,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '900 2100 MHz HSDPA' }
 - {tech: 4G, bands: '800 1800 2100 2600 MHz LTE' }
-peripherals: [Dual SIM, Accelerometer, Gyroscope, Light sensor, Proximity sensor, Compass, Hall sensor, GPS, FM Radio, USB OTG]
+peripherals: [Dual SIM, Accelerometer, Gyroscope, Light sensor, Proximity sensor, Compass, Hall effect sensor, GPS, FM radio, USB OTG]
 ram: 2/3 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Up</kbd> + <kbd>Power</kbd>.
 release: 2016
diff --git a/_data/devices/h811.yml b/_data/devices/h811.yml
index a9e801e..177d5ce 100644
--- a/_data/devices/h811.yml
+++ b/_data/devices/h811.yml
@@ -25,8 +25,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz' }
 - {tech: 3G, bands: '850 1700/2100 1900 2100 MHz UMTS' }
 - {tech: 4G, bands: '5(850) 4(1700/2100) 7(2600) 3(1800) 2(1900) 12(700)' }
-peripherals: [Accelerometer, Gyroscope, Proximity sensor, Compass, Barometer, Color spectrum,
-  FM Radio]
+peripherals: [Accelerometer, Gyroscope, Proximity sensor, Compass, Barometer, Color spectrum sensor, FM radio]
 ram: 3 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> + <kbd>Power</kbd> until the LG logo appears, then release <kbd>Power</kbd> for a second and hold it again until the recovery comes up.
 release: 2015-06
diff --git a/_data/devices/h815.yml b/_data/devices/h815.yml
index 865d96c..4bd75dd 100644
--- a/_data/devices/h815.yml
+++ b/_data/devices/h815.yml
@@ -25,7 +25,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '850 900 1900 2100 MHz HSDPA' }
 - {tech: 4G, bands: '1(2100) 17(700) 5(850) 4(1700/2100) 3(1800) 7(2600) 2(1900) 8(900) 20(800) 28(700)' }
-peripherals: [Accelerometer, Gyroscope, Proximity sensor, Compass, Barometer, Color spectrum, FM Radio]
+peripherals: [Accelerometer, Gyroscope, Proximity sensor, Compass, Barometer, Color spectrum sensor, FM radio]
 ram: 3 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> + <kbd>Power</kbd> until the LG logo appears, then release <kbd>Power</kbd> for a second and hold it again until the recovery comes up.
 release: 2015-06
diff --git a/_data/devices/h830.yml b/_data/devices/h830.yml
index 70b88eb..724eaec 100644
--- a/_data/devices/h830.yml
+++ b/_data/devices/h830.yml
@@ -26,7 +26,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '850 900 1900 2100 MHz HSDPA' }
 - {tech: 4G, bands: '1(2100) 17(700) 5(850) 4(1700/2100) 3(1800) 7(2600) 2(1900) 8(900) 20(800) 28(700)' }
-peripherals: [Fingerprint reader, Accelerometer, Gyroscope, Proximity sensor, Compass, Barometer, Color spectrum]
+peripherals: [Fingerprint reader, Accelerometer, Gyroscope, Proximity sensor, Compass, Barometer, Color spectrum sensor]
 ram: 4 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> + <kbd>Power</kbd> until the LG logo appears, then release <kbd>Power</kbd> for a second and hold it again until the recovery comes up.
 release: 2016-02
diff --git a/_data/devices/h850.yml b/_data/devices/h850.yml
index 4f7f45e..ae08d6c 100644
--- a/_data/devices/h850.yml
+++ b/_data/devices/h850.yml
@@ -26,7 +26,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '850 900 1900 2100 MHz HSDPA' }
 - {tech: 4G, bands: '1(2100) 17(700) 5(850) 4(1700/2100) 3(1800) 7(2600) 2(1900) 8(900) 20(800) 28(700)' }
-peripherals: [Fingerprint reader, Accelerometer, Gyroscope, Proximity sensor, Compass, Barometer, Color spectrum, FM Radio]
+peripherals: [Fingerprint reader, Accelerometer, Gyroscope, Proximity sensor, Compass, Barometer, Color spectrum sensor, FM radio]
 ram: 4 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> + <kbd>Power</kbd> until the LG logo appears, then release <kbd>Power</kbd> for a second and hold it again until the recovery comes up.
 release: 2016-02
diff --git a/_data/devices/h870.yml b/_data/devices/h870.yml
index d691d8c..0b0a4ab 100644
--- a/_data/devices/h870.yml
+++ b/_data/devices/h870.yml
@@ -26,7 +26,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '850 900 1900 2100 MHz HSDPA' }
 - {tech: 4G, bands: '1(2100) 17(700) 5(850) 4(1700/2100) 3(1800) 7(2600) 2(1900) 8(900) 20(800) 28(700)' }
-peripherals: [Fingerprint reader, Accelerometer, Barometer, Gyroscope, Proximity sensor, Compass, Barometer, Color spectrum]
+peripherals: [Fingerprint reader, Accelerometer, Barometer, Gyroscope, Proximity sensor, Compass, Color spectrum sensor]
 ram: 4 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> + <kbd>Power</kbd> until the LG logo appears, then release <kbd>Power</kbd> for a second and hold it again until the recovery comes up.
 release: 2017-02
diff --git a/_data/devices/h910.yml b/_data/devices/h910.yml
index d469b40..d2a1ade 100644
--- a/_data/devices/h910.yml
+++ b/_data/devices/h910.yml
@@ -26,7 +26,8 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '850 900 1900 2100 MHz HSDPA' }
 - {tech: 4G, bands: '1(2100) 17(700) 5(850) 4(1700/2100) 3(1800) 7(2600) 2(1900) 8(900) 20(800) 28(700)' }
-peripherals: [Fingerprint reader, Accelerometer, Gyroscope, Proximity sensor, Compass, Barometer, Color spectrum, Ambient light sensor, GPS, IR blaster, Infrared sensor, Mobility DisplayPort (MyDP), NFC, USB OTG]
+peripherals: [Fingerprint reader, Accelerometer, Gyroscope, Proximity sensor, Compass, Barometer, Color spectrum sensor,
+  Light sensor, GPS, IR blaster, Infrared sensor, Mobility DisplayPort (MyDP), NFC, USB OTG]
 ram: 4 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> + <kbd>Power</kbd> until the LG logo appears, then release <kbd>Power</kbd> for a second and hold it again until the recovery comes up.
 release: 2016-10
diff --git a/_data/devices/h918.yml b/_data/devices/h918.yml
index 61fdd7c..9f12f32 100644
--- a/_data/devices/h918.yml
+++ b/_data/devices/h918.yml
@@ -26,7 +26,8 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '850 900 1900 2100 MHz HSDPA' }
 - {tech: 4G, bands: '1(2100) 17(700) 5(850) 4(1700/2100) 3(1800) 7(2600) 2(1900) 8(900) 20(800) 28(700)' }
-peripherals: [Fingerprint reader, Accelerometer, Gyroscope, Proximity sensor, Compass, Barometer, Color spectrum, Ambient light sensor, GPS, IR blaster, Infrared sensor, Mobility DisplayPort (MyDP), NFC, USB OTG]
+peripherals: [Fingerprint reader, Accelerometer, Gyroscope, Proximity sensor, Compass, Barometer, Color spectrum sensor,
+  Light sensor, GPS, IR blaster, Infrared sensor, Mobility DisplayPort (MyDP), NFC, USB OTG]
 ram: 4 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> + <kbd>Power</kbd> until the LG logo appears, then release <kbd>Power</kbd> for a second and hold it again until the recovery comes up.
 release: 2016-10
diff --git a/_data/devices/hammerhead.yml b/_data/devices/hammerhead.yml
index f84411e..597e491 100644
--- a/_data/devices/hammerhead.yml
+++ b/_data/devices/hammerhead.yml
@@ -27,7 +27,7 @@ network:
 - {tech: 3G, bands: '800 850 900 1700 1900 2100 MHz HSDPA (North Amera model), 850 900 1700 1900 2100 MHz HSDPA' }
 - {tech: 4G, bands: '1(2100) 2(1900) 4(1700/2100) 5(850) 17(700) 19(800) 25(1900) 26(850) 41(2500) (North America), 1(2100) 3(1800) 5(850) 7(2600) 8(900) 20(800)' }
 no_oem_unlock_switch: true
-peripherals: ['']
+peripherals: None
 ram: 2 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> + <kbd>Volume Up</kbd>.
   Then press and hold <kbd>Power</kbd> until a boot menu appears.
diff --git a/_data/devices/hero2lte.yml b/_data/devices/hero2lte.yml
index 4be61b4..b4df86d 100644
--- a/_data/devices/hero2lte.yml
+++ b/_data/devices/hero2lte.yml
@@ -26,7 +26,8 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '850 900 1700(AWS) 1900 2100 MHz HSDPA' }
 - {tech: 4G, bands: 'B1-B8, B12-B13, B17-B20, B25-B26, B28, B38(TDD), B39-41(TDD)' }
-peripherals: [Fingerprint reader, Heart rate sensor, Motion Coprocessor, Accelerometer, Gesture sensor, Gyroscope, Proximity sensor, Compass, Barometer, Hall sensor, RGB ambient light]
+peripherals: [Fingerprint reader, Heart rate sensor, Motion Coprocessor, Accelerometer, Gesture sensor,
+  Gyroscope, Proximity sensor, Compass, Barometer, Hall effect sensor, RGB ambient light]
 ram: 4 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Up</kbd> + <kbd>Home</kbd> + <kbd>Power</kbd>. When the blue text appears, release the buttons.
 release: 2016-03-18
diff --git a/_data/devices/herolte.yml b/_data/devices/herolte.yml
index d466eb4..c8e777c 100644
--- a/_data/devices/herolte.yml
+++ b/_data/devices/herolte.yml
@@ -27,8 +27,8 @@ network:
 - {tech: 3G, bands: '850 900 1700(AWS) 1900 2100 MHz HSDPA' }
 - {tech: 4G, bands: 'B1-B8, B12-B13, B17-B20, B25-B26, B28, B38(TDD), B39-41(TDD)' }
 peripherals: [Fingerprint reader, Heart rate sensor, Motion Coprocessor, Accelerometer,
-  Gesture sensor, Gyroscope, Proximity sensor, Compass, Barometer, Hall sensor,
-  Hall sensor, RGB ambient light, Infrared sensor]
+  Gesture sensor, Gyroscope, Proximity sensor, Compass, Barometer, Hall effect sensor,
+  RGB ambient light, Infrared sensor]
 ram: 4 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Up</kbd> + <kbd>Home</kbd> + <kbd>Power</kbd>. When the blue text appears, release the buttons.
 release: 2016-03-18
diff --git a/_data/devices/hiaeuhl.yml b/_data/devices/hiaeuhl.yml
index f51064d..058dd1f 100644
--- a/_data/devices/hiaeuhl.yml
+++ b/_data/devices/hiaeuhl.yml
@@ -26,7 +26,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '850 900 1900 2100 MHz HSDPA' }
 - {tech: 4G, bands: '1(2100) 3(1800) 5(850) 7(2600) 8(900) 20(800) 28(700) 38(2600) 40(2300) 41(2500) MHz LTE' }
-peripherals: [Fingerprint reader, Accelerometer, Gyroscope, Proximity sensor, Compass, Magnetic sensor, GPS, NFC]
+peripherals: [Fingerprint reader, Accelerometer, Gyroscope, Proximity sensor, Hall effect sensor, Compass, GPS, NFC]
 ram: 2/3 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> + <kbd>Power</kbd> until the bootloader appears, then release the buttons.
   Use the volume buttons to scroll to "RECOVERY" and press <kbd>Power</kbd> to select.
diff --git a/_data/devices/hiaeul.yml b/_data/devices/hiaeul.yml
index 2ade418..f9a1c9b 100644
--- a/_data/devices/hiaeul.yml
+++ b/_data/devices/hiaeul.yml
@@ -26,7 +26,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '850 1700(AWS) 1900 2100 MHz HSDPA' }
 - {tech: 4G, bands: '2(1900) 3(1800) 4(1700) 5(850) 7(2600) 12(700) 13(700) 17(700) 19(850) MHz LTE' }
-peripherals: [Fingerprint reader, Accelerometer, Gyroscope, Proximity sensor, Compass, Magnetic sensor, GPS, NFC]
+peripherals: [Fingerprint reader, Accelerometer, Gyroscope, Proximity sensor, Hall effect sensor, Compass, GPS, NFC]
 ram: 2/3 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> + <kbd>Power</kbd> until the bootloader appears, then release the buttons.
   Use the volume buttons to scroll to "RECOVERY" and press <kbd>Power</kbd> to select.
diff --git a/_data/devices/himaul.yml b/_data/devices/himaul.yml
index 31fea80..435d45e 100644
--- a/_data/devices/himaul.yml
+++ b/_data/devices/himaul.yml
@@ -26,7 +26,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '8850 1900 2100 MHz HSDPA (AT&T),850 1700(AWS) 1900 2100 MHz HSDPA (T-Mobile)' }
 - {tech: 4G, bands: '700 1700 1800 1900 2600 MHz (AT&T), 700 850 1800 1900 2100 2600 MHz (T-Mobile)' }
-peripherals: [Dual front speakers, Accelerometer, Gyroscope, Proximity sensor, Compass, GPS, Hall sensor, NFC, IR blaster]
+peripherals: [Dual front speakers, Accelerometer, Gyroscope, Proximity sensor, Compass, GPS, Hall effect sensor, NFC, IR blaster]
 ram: 3 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> + <kbd>Power</kbd> until the bootloader appears, then release the buttons.
   Use the volume buttons to scroll to "RECOVERY" and press <kbd>Power</kbd> to select.
diff --git a/_data/devices/himawl.yml b/_data/devices/himawl.yml
index 6774199..e4de971 100644
--- a/_data/devices/himawl.yml
+++ b/_data/devices/himawl.yml
@@ -26,7 +26,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM, 800 1900 MHz CDMA' }
 - {tech: 3G, bands: '850 900 1900 2100 MHz HSDPA, CDMA2000 1xEV-DO' }
 - {tech: 4G, bands: '4(1700/2100), 13(700)' }
-peripherals: [Dual front speakers, Accelerometer, Gyroscope, Proximity sensor, Compass, GPS, Hall sensor, NFC, IR blaster]
+peripherals: [Dual front speakers, Accelerometer, Gyroscope, Proximity sensor, Compass, GPS, Hall effect sensor, NFC, IR blaster]
 ram: 3 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> + <kbd>Power</kbd> until the bootloader appears, then release the buttons.
   Use the volume buttons to scroll to "RECOVERY" and press <kbd>Power</kbd> to select.
diff --git a/_data/devices/hydrogen.yml b/_data/devices/hydrogen.yml
index 6ae414a..185c0f9 100644
--- a/_data/devices/hydrogen.yml
+++ b/_data/devices/hydrogen.yml
@@ -26,7 +26,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM, 800 1900 MHz CDMA' }
 - {tech: 3G, bands: '850 900 1900 2100 MHz HSDPA, CDMA2000 1xEV-DO TD-SCDMA' }
 - {tech: 4G, bands: '1(2100) 3(1800) 7(2600) 38(2600) 39(1900) 40(2300) 41(2500)' }
-peripherals: [Fingerprint reader, Accelerometer, Gyroscope, Proximity sensor, Compass, GPS, Infrared sensor, FM Radio]
+peripherals: [Fingerprint reader, Accelerometer, Gyroscope, Proximity sensor, Compass, GPS, Infrared sensor, FM radio]
 ram: 3 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Up</kbd> + <kbd>Power</kbd>.
   Keep holding both buttons until the "MI" logo appears on the screen, then release.
diff --git a/_data/devices/i9100.yml b/_data/devices/i9100.yml
index 59aaff6..1dbfe41 100644
--- a/_data/devices/i9100.yml
+++ b/_data/devices/i9100.yml
@@ -26,7 +26,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM'}
 - {tech: 3G, bands: '850 900 1900 2100 MHz HSPA+'}
 - {tech: 4G, bands: ''}
-peripherals: ['']
+peripherals: None
 ram: 1 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Up</kbd> + <kbd>Home</kbd> + <kbd>Power</kbd>.
 release: 2011-02-11
diff --git a/_data/devices/i9305.yml b/_data/devices/i9305.yml
index 302c486..6831f41 100644
--- a/_data/devices/i9305.yml
+++ b/_data/devices/i9305.yml
@@ -26,7 +26,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM'}
 - {tech: 3G, bands: '850 900 2100 MHz UMTS/HSPA+'}
 - {tech: 4G, bands: '800 1800 2600 MHz (GT-I9305), 900 1800 2600 MHz (GT-I9305N), 1800 2600 MHz (GT-I9305T)'}
-peripherals: [Accelerometer, Gyroscope, Proximity sensor, Digital compass, GPS, NFC]
+peripherals: [Accelerometer, Gyroscope, Proximity sensor, Compass, GPS, NFC]
 ram: 2 GB
 recovery_boot: With the device powered off, hold <kbd>Home</kbd> + <kbd>Volume Up</kbd> + <kbd>Power</kbd>.
 release: 2012-10
diff --git a/_data/devices/ido.yml b/_data/devices/ido.yml
index a887eca..c78b917 100644
--- a/_data/devices/ido.yml
+++ b/_data/devices/ido.yml
@@ -26,7 +26,9 @@ network:
 - {tech: 2G, bands: '900 1800 1900 MHz GSM, 800 1900 MHz CDMA' }
 - {tech: 3G, bands: '850 900 1900 2100 MHz HSDPA' }
 - {tech: 4G, bands: '1(2100) 3(1800) 7(2600) 38(2600) 39(1900) 40(2300) 41(2500)' }
-peripherals: [Dual SIM, Fingerprint reader (Prime only), Accelerometer, Gyroscope, Proximity sensor, Compass, IR blaster, FM Radio, GPS]
+peripherals:
+- All: [Dual SIM, Accelerometer, Gyroscope, Proximity sensor, Compass, IR blaster, FM radio, GPS]
+- Prime: [Fingerprint reader]
 ram: 2/3 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Up</kbd> + <kbd>Power</kbd>.
   Keep holding both buttons until the "MI" logo appears on the screen, then release.
diff --git a/_data/devices/ivy.yml b/_data/devices/ivy.yml
index 374a317..bda8864 100644
--- a/_data/devices/ivy.yml
+++ b/_data/devices/ivy.yml
@@ -25,7 +25,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '850 900 1900 2100 MHz HSDPA' }
 - {tech: 4G, bands: '1 (2100) 2 (1900) 3 (1800) 4 (1700/2100) 5 (850) 7 (2600) 8 (900) 12 (700) 17 (700) 20 (800) 28 (700) 40 (2300) LTE' }
-peripherals: [Accelerometer, Gyroscope, Proximity sensor, Ambient light, GPS]
+peripherals: [Accelerometer, Gyroscope, Proximity sensor, Light sensor, GPS]
 ram: 3 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> + <kbd>Power</kbd>.
 release: 2015-06
diff --git a/_data/devices/jag3gds.yml b/_data/devices/jag3gds.yml
index 055c179..0eecf33 100644
--- a/_data/devices/jag3gds.yml
+++ b/_data/devices/jag3gds.yml
@@ -25,7 +25,7 @@ network:
 - {tech: 2G, bands: '900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '900 2100 MHz HSDPA' }
 - {tech: 4G, bands: '' }
-peripherals: [GPS, Accelerometer, Compass, Proximity sensor, FM Radio]
+peripherals: [GPS, Accelerometer, Compass, Proximity sensor, FM radio]
 ram: 1 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> + <kbd>Power</kbd>.
 recovery_partition: /dev/block/platform/msm_sdcc.1/by-name/recovery
diff --git a/_data/devices/jagnm.yml b/_data/devices/jagnm.yml
index 1eb693c..c1ce7a8 100644
--- a/_data/devices/jagnm.yml
+++ b/_data/devices/jagnm.yml
@@ -25,7 +25,7 @@ network:
 - {tech: 2G, bands: '900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '900 2100 MHz HSDPA' }
 - {tech: 4G, bands: '1(2100) 3(1800) 7(2600) 8(900) 20(800) LTE' }
-peripherals: [NFC, GPS, Accelerometer, Compass, Proximity sensor, FM Radio]
+peripherals: [NFC, GPS, Accelerometer, Compass, Proximity sensor, FM radio]
 ram: 1 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> + <kbd>Power</kbd>.
 recovery_partition: /dev/block/platform/msm_sdcc.1/by-name/recovery
diff --git a/_data/devices/jfltevzw.yml b/_data/devices/jfltevzw.yml
index d4e6244..5344a2b 100644
--- a/_data/devices/jfltevzw.yml
+++ b/_data/devices/jfltevzw.yml
@@ -27,7 +27,8 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM, 800 1900 MHz CDMA' }
 - {tech: 3G, bands: '850 900 1900 2100 MHz HSPA+ / UMTS, CDMA2000 1xEV-DO' }
 - {tech: 4G, bands: '1700/2100 700 MHz LTE' }
-peripherals: [GPS, GLONASS, NFC, Infrared sensor, USB OTG, MHL 2.0, HDMI (TV-out via MHL A\V link), Accelerometer, Barometer, Gesture sensor, Gyroscope, Hall sensor, Hygrometer, Proximity sensor, RGB light sensor, Thermometer]
+peripherals: [GPS, GLONASS, NFC, Infrared sensor, USB OTG, MHL 2.0, MHL HDMI, Accelerometer, Barometer,
+  Gesture sensor, Gyroscope, Hall effect sensor, Hygrometer, Proximity sensor, Light sensor, Thermometer]
 ram: 2 GB
 recovery_boot: With the device powered off, hold <kbd>Home</kbd> + <kbd>Volume Up</kbd> + <kbd>Power</kbd>.
 release: 2013-04
diff --git a/_data/devices/jfltexx.yml b/_data/devices/jfltexx.yml
index 2324521..f129933 100644
--- a/_data/devices/jfltexx.yml
+++ b/_data/devices/jfltexx.yml
@@ -26,7 +26,8 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '850 900 1900 2100 MHz HSPA+' }
 - {tech: 4G, bands: '800 850 900 1800 2100 2600 MHz LTE' }
-peripherals: [GPS, GLONASS, NFC, Infrared sensor, USB OTG, MHL 2.0, HDMI (TV-out via MHL A\V link), Accelerometer, Barometer, Gesture sensor, Gyroscope, Hall sensor, Hygrometer, Proximity sensor, RGB light sensor, Thermometer]
+peripherals: [GPS, GLONASS, NFC, Infrared sensor, USB OTG, MHL 2.0, MHL HDMI, Accelerometer, Barometer,
+  Gesture sensor, Gyroscope, Hall effect sensor, Hygrometer, Proximity sensor, Light sensor, Thermometer]
 ram: 2 GB
 recovery_boot: With the device powered off, hold <kbd>Home</kbd> + <kbd>Volume Up</kbd> + <kbd>Power</kbd>.
 release: 2013-04
diff --git a/_data/devices/k3gxx.yml b/_data/devices/k3gxx.yml
index b8c430f..a054075 100644
--- a/_data/devices/k3gxx.yml
+++ b/_data/devices/k3gxx.yml
@@ -26,7 +26,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '850 900 1900 2100 MHz HSDPA' }
 - {tech: 4G, bands: '' }
-peripherals: [Accelerometer, Gyroscope, Proximity sensor, Compass, Barometer, Temperature, Humidity, NFC, GPS, GLONASS, Fingerprint reader]
+peripherals: [Accelerometer, Gyroscope, Proximity sensor, Compass, Barometer, Thermometer, Hygrometer, NFC, GPS, GLONASS, Fingerprint reader]
 ram: 2 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Up</kbd> + <kbd>Home</kbd> + <kbd>Power</kbd>.
 release: 2014-03
diff --git a/_data/devices/karin.yml b/_data/devices/karin.yml
index 39f430a..84e92ca 100644
--- a/_data/devices/karin.yml
+++ b/_data/devices/karin.yml
@@ -26,7 +26,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '850 900 1900 2100 MHz HSDPA' }
 - {tech: 4G, bands: '1 (2100) 2 (1900) 3 (1800) 4 (1700/2100) 5 (850) 7 (2600) 8 (900) 12 (700) 17 (700) 20 (800) 28 (700) 40 (2300) LTE' }
-peripherals: [Accelerometer, Gyroscope, Proximity sensor, Ambient light sensor, GPS]
+peripherals: [Accelerometer, Gyroscope, Proximity sensor, Light sensor, GPS]
 ram: 3 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> + <kbd>Power</kbd>.
 release: 2015-10
diff --git a/_data/devices/karin_windy.yml b/_data/devices/karin_windy.yml
index 73c270a..8924514 100644
--- a/_data/devices/karin_windy.yml
+++ b/_data/devices/karin_windy.yml
@@ -26,7 +26,7 @@ network:
 - {tech: 2G, bands: '' }
 - {tech: 3G, bands: '' }
 - {tech: 4G, bands: '' }
-peripherals: [Accelerometer, Gyroscope, Proximity sensor, Ambient light sensor, GPS]
+peripherals: [Accelerometer, Gyroscope, Proximity sensor, Light sensor, GPS]
 ram: 3 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> + <kbd>Power</kbd>.
 release: 2015-10
diff --git a/_data/devices/kccat6.yml b/_data/devices/kccat6.yml
index f8cb678..0bab6ce 100644
--- a/_data/devices/kccat6.yml
+++ b/_data/devices/kccat6.yml
@@ -25,7 +25,8 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '850 900 1900 2100 MHz HSDPA' }
 - {tech: 4G, bands: '1(2100) 2(1900) 3(1800) 5(850) 7(2600) 8(900) 20(800)' }
-peripherals: [Fingerprint reader, Heart rate sensor, Motion Coprocessor, Accelerometer, Gesture sensor, Gyroscope, Proximity sensor, Compass, Barometer, Hall sensor, RGB ambient light, Infrared sensor]
+peripherals: [Fingerprint reader, Heart rate sensor, Motion Coprocessor, Accelerometer, Gesture sensor,
+  Gyroscope, Proximity sensor, Compass, Barometer, Hall effect sensor, RGB ambient light, Infrared sensor]
 ram: 2 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Up</kbd> + <kbd>Home</kbd> + <kbd>Power</kbd>. When the blue text appears, release the buttons.
 release: 2014-08-21
diff --git a/_data/devices/kenzo.yml b/_data/devices/kenzo.yml
index 02658ae..13379ab 100644
--- a/_data/devices/kenzo.yml
+++ b/_data/devices/kenzo.yml
@@ -26,7 +26,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '850 900 1700(AWS) 1900 2100 MHz HSDPA, 850 900 1900 2100 MHz HSDPA (India)' }
 - {tech: 4G, bands: '1(2100) 2(1900) 3(1800) 4(1700/2100) 5(850) 7(2600) 8(900) 38(2600) 39(1900) 40(2300) 41(2500), 1(2100) 3(1800) 5(850) 7(2600) 8(900) 38(2600) 39(1900) 40(2300) 41(2500) (India)' }
-peripherals: [Fingerprint reader, Accelerometer, Gyroscope, Proximity sensor, Compass, GPS, Infrared sensor, FM Radio]
+peripherals: [Fingerprint reader, Accelerometer, Gyroscope, Proximity sensor, Compass, GPS, Infrared sensor, FM radio]
 ram: 2/3 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Up</kbd> + <kbd>Power</kbd>.
   Keep holding both buttons until the "MI" logo appears on the screen, then release.
diff --git a/_data/devices/kingdom.yml b/_data/devices/kingdom.yml
index c246b50..44d652c 100644
--- a/_data/devices/kingdom.yml
+++ b/_data/devices/kingdom.yml
@@ -27,7 +27,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '850 900 1700 1900 2100 MHz HSDPA' }
 - {tech: 4G, bands: '1(2100) 3(1800) 5(850) 7(2600) 8(900) 20(800)' }
-peripherals: [GPS, NFC, FM Radio, Hall sensor, Light sensor, Proximity sensor, Accelerometer, Gyroscope, Compass, RGB ambient light]
+peripherals: [GPS, NFC, FM radio, Hall effect sensor, Light sensor, Proximity sensor, Accelerometer, Gyroscope, Compass, RGB ambient light]
 ram: 3 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> + <kbd>Volume Up</kbd>.
   Then press and hold the <kbd>Power</kbd> button until the "Lenovo" logo appears.
diff --git a/_data/devices/kipper.yml b/_data/devices/kipper.yml
index ab601ba..e278b64 100644
--- a/_data/devices/kipper.yml
+++ b/_data/devices/kipper.yml
@@ -25,7 +25,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '850 900 1900 2100 MHz HSDPA' }
 - {tech: 4G, bands: '800 900 1800 2100 2600 MHz' }
-peripherals: [Dual SIM, FM Radio, Accelerometer, Gyroscope, Proximity sensor, Compass]
+peripherals: [Dual SIM, FM radio, Accelerometer, Gyroscope, Proximity sensor, Compass]
 ram: 3 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> + <kbd>Power</kbd>
 release: 2015-11
diff --git a/_data/devices/kiwi.yml b/_data/devices/kiwi.yml
index f095340..55ff257 100644
--- a/_data/devices/kiwi.yml
+++ b/_data/devices/kiwi.yml
@@ -26,7 +26,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '850 1700 (AWS) 1900 2100 MHz HSDPA (USA models), 850 900 1900 2100 MHz HSDPA (China models)' }
 - {tech: 4G, bands: '1(2100) 3(1800) 7(2600) 8(900) 20(800) 2(1900) 4(1700/2100) 5(850) 7(2600) 12(700) 17(700) LTE (USA models), 1(2100) 3(1800) 38(2600) 39(1900) 40(2300) 41(2500) LTE (China models)' }
-peripherals: [Fingerprint reader, Accelerometer, Proximity sensor, Light sensor, Compass, GPS, Hall sensor]
+peripherals: [Fingerprint reader, Accelerometer, Proximity sensor, Light sensor, Compass, GPS, Hall effect sensor]
 ram: 2 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Up</kbd> + <kbd>Power</kbd>.
   Keep holding buttons until the "Honor" logo appears, then release all buttons.
diff --git a/_data/devices/klimtwifi.yml b/_data/devices/klimtwifi.yml
index 7c66797..78eb25b 100644
--- a/_data/devices/klimtwifi.yml
+++ b/_data/devices/klimtwifi.yml
@@ -26,7 +26,9 @@ network:
 - {tech: 2G, bands: '' }
 - {tech: 3G, bands: '' }
 - {tech: 4G, bands: '' }
-peripherals: [Accelerometer, Compass, GPS, Gyroscope, Fingerprint reader (not supported)]
+peripherals:
+- SM-T700: [Accelerometer, Compass, GPS, Gyroscope]
+- Unsupported: [Fingerprint reader]
 ram: 3 GB
 recovery_boot: With the device powered off, hold <kbd>Home</kbd> + <kbd>Volume Up</kbd> + <kbd>Power</kbd>.
 release: 2014
diff --git a/_data/devices/klte.yml b/_data/devices/klte.yml
index 8c78965..d4f4c4f 100644
--- a/_data/devices/klte.yml
+++ b/_data/devices/klte.yml
@@ -24,7 +24,8 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM, 800 1900MHz 1xRTT (G900R4/G900R7/G900V/S902L)' }
 - {tech: 3G, bands: '850 900 1900 2100 MHz HSDPA (G900F/G900V), 850 900 1700 1900 2100 MHz HSDPA (G900M), 850 1700 1900 2100 MHz HSDPA (G900T), 800 1900MHz EVDO (G900R4/G900R7/G900V/S902L)' }
 - {tech: 4G, bands: '1(2100) 2(1900) 3(1800) 5(850) 7(2600) 8(900) 20(800) (G900F), 1(2100) 2(1900) 4(1700/2100) 5(850) 7(2600) 17(700) (G900M), 1(2100) 2(1900) 3(1800) 4(1700/2100) 5(850) 7(2600) 8(900) 17(700) (G900T), 2(1900) 4(1700/2100) 5(850) 12(700) 17(700) (G900R4), 2(1900) 4(1700/2100) 12(700) (G900R7), 4(1700/2100) 13(700) (G900V/S902L)' }
-peripherals: [Fingerprint reader, Heart rate sensor, Motion Coprocessor, Accelerometer, Gesture sensor, Gyroscope, Proximity sensor, Compass, Barometer, Hall sensor, RGB ambient light, Infrared sensor]
+peripherals: [Fingerprint reader, Heart rate sensor, Motion Coprocessor, Accelerometer, Gesture sensor,
+  Gyroscope, Proximity sensor, Compass, Barometer, Hall effect sensor, RGB ambient light, Infrared sensor]
 ram: 2 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Up</kbd> + <kbd>Home</kbd> + <kbd>Power</kbd>. When the blue text appears, release the buttons.
 release: 2014-04-11
diff --git a/_data/devices/kltechn.yml b/_data/devices/kltechn.yml
index e521995..34e9ba4 100644
--- a/_data/devices/kltechn.yml
+++ b/_data/devices/kltechn.yml
@@ -24,7 +24,8 @@ network:
 - {tech: 2G, bands: '' }
 - {tech: 3G, bands: '' }
 - {tech: 4G, bands: '' }
-peripherals: [Fingerprint reader, Heart rate sensor, Motion Coprocessor, Accelerometer, Gesture sensor, Gyroscope, Proximity sensor, Compass, Barometer, Hall sensor, RGB ambient light, Infrared sensor]
+peripherals: [Fingerprint reader, Heart rate sensor, Motion Coprocessor, Accelerometer, Gesture sensor,
+  Gyroscope, Proximity sensor, Compass, Barometer, Hall effect sensor, RGB ambient light, Infrared sensor]
 ram: 2 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Up</kbd> + <kbd>Home</kbd> + <kbd>Power</kbd>. When the blue text appears, release the buttons.
 release:
diff --git a/_data/devices/kltechnduo.yml b/_data/devices/kltechnduo.yml
index 9bd22bd..64fd5f3 100644
--- a/_data/devices/kltechnduo.yml
+++ b/_data/devices/kltechnduo.yml
@@ -24,7 +24,8 @@ network:
 - {tech: 2G, bands: '' }
 - {tech: 3G, bands: '' }
 - {tech: 4G, bands: '' }
-peripherals: [Fingerprint reader, Heart rate sensor, Motion Coprocessor, Accelerometer, Gesture sensor, Gyroscope, Proximity sensor, Compass, Barometer, Hall sensor, RGB ambient light, Infrared sensor]
+peripherals: [Fingerprint reader, Heart rate sensor, Motion Coprocessor, Accelerometer, Gesture sensor,
+  Gyroscope, Proximity sensor, Compass, Barometer, Hall effect sensor, RGB ambient light, Infrared sensor]
 ram: 2 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Up</kbd> + <kbd>Home</kbd> + <kbd>Power</kbd>. When the blue text appears, release the buttons.
 release:
diff --git a/_data/devices/klteduos.yml b/_data/devices/klteduos.yml
index 15198ea..109557a 100644
--- a/_data/devices/klteduos.yml
+++ b/_data/devices/klteduos.yml
@@ -24,7 +24,8 @@ network:
 - {tech: 2G, bands: '' }
 - {tech: 3G, bands: '' }
 - {tech: 4G, bands: '' }
-peripherals: [Fingerprint reader, Heart rate sensor, Motion Coprocessor, Accelerometer, Gesture sensor, Gyroscope, Proximity sensor, Compass, Barometer, Hall sensor, RGB ambient light, Infrared sensor]
+peripherals: [Fingerprint reader, Heart rate sensor, Motion Coprocessor, Accelerometer, Gesture sensor,
+  Gyroscope, Proximity sensor, Compass, Barometer, Hall effect sensor, RGB ambient light, Infrared sensor]
 ram: 2 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Up</kbd> + <kbd>Home</kbd> + <kbd>Power</kbd>. When the blue text appears, release the buttons.
 release:
diff --git a/_data/devices/kltedv.yml b/_data/devices/kltedv.yml
index 2e1382e..e81189d 100644
--- a/_data/devices/kltedv.yml
+++ b/_data/devices/kltedv.yml
@@ -24,7 +24,8 @@ network:
 - {tech: 2G, bands: '' }
 - {tech: 3G, bands: '' }
 - {tech: 4G, bands: '' }
-peripherals: [Fingerprint reader, Heart rate sensor, Motion Coprocessor, Accelerometer, Gesture sensor, Gyroscope, Proximity sensor, Compass, Barometer, Hall sensor, RGB ambient light, Infrared sensor]
+peripherals: [Fingerprint reader, Heart rate sensor, Motion Coprocessor, Accelerometer, Gesture sensor,
+  Gyroscope, Proximity sensor, Compass, Barometer, Hall effect sensor, RGB ambient light, Infrared sensor]
 ram: 2 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Up</kbd> + <kbd>Home</kbd> + <kbd>Power</kbd>. When the blue text appears, release the buttons.
 release: 2014-04
diff --git a/_data/devices/kltekdi.yml b/_data/devices/kltekdi.yml
index 50cf619..fb38d84 100644
--- a/_data/devices/kltekdi.yml
+++ b/_data/devices/kltekdi.yml
@@ -24,7 +24,8 @@ network:
 - {tech: 2G, bands: '' }
 - {tech: 3G, bands: '' }
 - {tech: 4G, bands: '' }
-peripherals: [Fingerprint reader, Heart rate sensor, Motion Coprocessor, Accelerometer, Gesture sensor, Gyroscope, Proximity sensor, Compass, Barometer, Hall sensor, RGB ambient light, Infrared sensor]
+peripherals: [Fingerprint reader, Heart rate sensor, Motion Coprocessor, Accelerometer, Gesture sensor,
+  Gyroscope, Proximity sensor, Compass, Barometer, Hall effect sensor, RGB ambient light, Infrared sensor]
 ram: 2 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Up</kbd> + <kbd>Home</kbd> + <kbd>Power</kbd>. When the blue text appears, release the buttons.
 release: 2014-05
diff --git a/_data/devices/kltekor.yml b/_data/devices/kltekor.yml
index 12a0ae0..9216e27 100644
--- a/_data/devices/kltekor.yml
+++ b/_data/devices/kltekor.yml
@@ -24,7 +24,8 @@ network:
 - {tech: 2G, bands: '' }
 - {tech: 3G, bands: '' }
 - {tech: 4G, bands: '' }
-peripherals: [Fingerprint reader, Heart rate sensor, Motion Coprocessor, Accelerometer, Gesture sensor, Gyroscope, Proximity sensor, Compass, Barometer, Hall sensor, RGB ambient light, Infrared sensor]
+peripherals: [Fingerprint reader, Heart rate sensor, Motion Coprocessor, Accelerometer, Gesture sensor, Gyroscope,
+  Proximity sensor, Compass, Barometer, Hall effect sensor, RGB ambient light, Infrared sensor]
 ram: 2 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Up</kbd> + <kbd>Home</kbd> + <kbd>Power</kbd>. When the blue text appears, release the buttons.
 release: 2014-04
diff --git a/_data/devices/ks01lte.yml b/_data/devices/ks01lte.yml
index ceeeefc..6afedbc 100644
--- a/_data/devices/ks01lte.yml
+++ b/_data/devices/ks01lte.yml
@@ -26,7 +26,8 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '850 900 1900 2100 MHz HSDPA' }
 - {tech: 4G, bands: '1(2100) 3(1800) 5(850) 7(2600) 8(900) 20(800) LTE' }
-peripherals: [GPS, GLONASS, NFC, Infrared sensor, USB OTG, MHL 2.0, HDMI (TV-out via MHL A\V link), Accelerometer, Barometer, Gesture sensor, Gyroscope, Hall sensor, Hygrometer, Proximity sensor, RGB light sensor, Thermometer]
+peripherals: [GPS, GLONASS, NFC, Infrared sensor, USB OTG, MHL 2.0, MHL HDMI, Accelerometer, Barometer, Gesture sensor,
+  Gyroscope, Hall effect sensor, Hygrometer, Proximity sensor, Light sensor, Thermometer]
 ram: 2 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Up</kbd> + <kbd>Home</kbd> + <kbd>Power</kbd>. When the blue text appears, release the buttons.
 release: 2013-10
diff --git a/_data/devices/land.yml b/_data/devices/land.yml
index 881f7af..dc389e9 100644
--- a/_data/devices/land.yml
+++ b/_data/devices/land.yml
@@ -26,7 +26,7 @@ network:
 - {tech: 2G, bands: '900 1800 1900 MHz GSM, 800 1900 MHz CDMA' }
 - {tech: 3G, bands: '850 900 1900 2100 MHz HSDPA' }
 - {tech: 4G, bands: '1(2100) 3(1800) 7(2600) 38(2600) 39(1900) 40(2300) 41(2500)' }
-peripherals: [Dual SIM, Fingerprint reader , Accelerometer, Gyroscope, Proximity sensor, Compass, IR blaster, FM Radio, GPS]
+peripherals: [Dual SIM, Fingerprint reader , Accelerometer, Gyroscope, Proximity sensor, Compass, IR blaster, FM radio, GPS]
 ram: 2/3 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Up</kbd> + <kbd>Volume Down</kbd> + <kbd>Power</kbd>.
 release: 2016-06
diff --git a/_data/devices/lentislte.yml b/_data/devices/lentislte.yml
index 9f003b7..a580870 100644
--- a/_data/devices/lentislte.yml
+++ b/_data/devices/lentislte.yml
@@ -25,7 +25,8 @@ network:
 - {tech: 2G, bands: '900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '850 1900 2100 MHz HSDPA' }
 - {tech: 4G, bands: '1(2100) 3(1800) 5(850) 7(2600) 17(700)' }
-peripherals: [Fingerprint reader, Heart rate sensor, Motion Coprocessor, Accelerometer, Gesture sensor, Gyroscope, Proximity sensor, Compass, Barometer, Hall sensor, RGB ambient light, IR blaster]
+peripherals: [Fingerprint reader, Heart rate sensor, Motion Coprocessor, Accelerometer, Gesture sensor, Gyroscope, Proximity sensor,
+  Compass, Barometer, Hall effect sensor, RGB ambient light, IR blaster]
 ram: 3 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Up</kbd> + <kbd>Home</kbd> + <kbd>Power</kbd>. When the blue text appears, release the buttons.
 release: 2014-07-15
diff --git a/_data/devices/libra.yml b/_data/devices/libra.yml
index 5a4bf88..e3f093c 100644
--- a/_data/devices/libra.yml
+++ b/_data/devices/libra.yml
@@ -26,7 +26,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 GSM' }
 - {tech: 3G, bands: '850 900 1800 2100 HSDPA' }
 - {tech: 4G, bands: '1(2100), 3(1800), 7(2600), 38(2600), 39(1900), 40(2300), 41(2500) LTE' }
-peripherals: [A-GPS, Accelerometer, Dual SIM, GLONASS, GPS, Gyroscope, Hall sensor, IR blaster, Light sensor, Proximity sensor, USB OTG]
+peripherals: [A-GPS, Accelerometer, Dual SIM, GLONASS, GPS, Gyroscope, Hall effect sensor, IR blaster, Light sensor, Proximity sensor, USB OTG]
 ram: 2/3 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Up</kbd> + <kbd>Power</kbd>.
   Keep holding both buttons until the "MI" logo appears on the screen, then release.
diff --git a/_data/devices/lithium.yml b/_data/devices/lithium.yml
index 9147a75..d2a3c27 100644
--- a/_data/devices/lithium.yml
+++ b/_data/devices/lithium.yml
@@ -26,7 +26,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM, 800 1900 MHz CDMA' }
 - {tech: 3G, bands: '850 900 1800 1900 MHz HSPA, 1900 2000 MHz TD-CDMA' }
 - {tech: 4G, bands: '850(B5) 1800(B3) 2100(B1) 2600(B7) MHz FDD-LTE, 1900(B39) 2300(B40) 2500(B41) 2600(B38) MHz TDD-LTE' }
-peripherals: [Dual SIM, Fingerprint reader, Accelerometer, Gyroscope, Light sensor, Ultrasound proximity sensor, Compass, NFC, GPS]
+peripherals: [Dual SIM, Fingerprint reader, Accelerometer, Gyroscope, Light sensor, Proximity sensor, Compass, NFC, GPS]
 ram: 4/6 GB LPDDR4
 recovery_boot: With the device powered off, hold <kbd>Volume Up</kbd> + <kbd>Power</kbd>.
   Keep holding both buttons until the "MI" logo appears on the screen, then release.
diff --git a/_data/devices/ls990.yml b/_data/devices/ls990.yml
index d19100c..ef0a677 100644
--- a/_data/devices/ls990.yml
+++ b/_data/devices/ls990.yml
@@ -25,7 +25,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM; 800 1900 MHz CDMA' }
 - {tech: 3G, bands: '850 1900 2100 MHz HSDPA' }
 - {tech: 4G, bands: '25(1900) 26(850) 41(2500) LTE' }
-peripherals: [Accelerometer, Compass, Gyroscope, Infrared sensor, Proximity sensor, FM Radio]
+peripherals: [Accelerometer, Compass, Gyroscope, Infrared sensor, Proximity sensor, FM radio]
 ram: 3 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> + <kbd>Power</kbd>.
 recovery_partition: /dev/block/platform/msm_sdcc.1/by-name/recovery
diff --git a/_data/devices/ls997.yml b/_data/devices/ls997.yml
index c2d9850..72a070a 100644
--- a/_data/devices/ls997.yml
+++ b/_data/devices/ls997.yml
@@ -26,7 +26,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '850 900 1900 2100 MHz HSDPA' }
 - {tech: 4G, bands: '1(2100) 17(700) 5(850) 4(1700/2100) 3(1800) 7(2600) 2(1900) 8(900) 20(800) 28(700)' }
-peripherals: [Fingerprint reader, Accelerometer, Gyroscope, Proximity sensor, Compass, Barometer, Color spectrum, Ambient light sensor, GPS, IR blaster, Infrared sensor, Mobility DisplayPort (MyDP), NFC, USB OTG]
+peripherals: [Fingerprint reader, Accelerometer, Gyroscope, Proximity sensor, Compass, Barometer, Color spectrum sensor, Light sensor, GPS, IR blaster, Infrared sensor, Mobility DisplayPort (MyDP), NFC, USB OTG]
 ram: 4 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> + <kbd>Power</kbd> until the LG logo appears, then release <kbd>Power</kbd> for a second and hold it again till recovery comes up.
 release: 2016-10
diff --git a/_data/devices/lt03lte.yml b/_data/devices/lt03lte.yml
index 3309f27..77bc2cf 100644
--- a/_data/devices/lt03lte.yml
+++ b/_data/devices/lt03lte.yml
@@ -23,7 +23,7 @@ network:
 - {tech: 2G, bands: '' }
 - {tech: 3G, bands: '' }
 - {tech: 4G, bands: '' }
-peripherals: ['']
+peripherals: None
 ram: 3 GB
 recovery_boot: With the device powered off, hold <kbd>Home</kbd> + <kbd>Volume Up</kbd> + <kbd>Power</kbd>.
 release: 2013-10
diff --git a/_data/devices/lux.yml b/_data/devices/lux.yml
index 7ecc674..c8f6f7c 100644
--- a/_data/devices/lux.yml
+++ b/_data/devices/lux.yml
@@ -27,7 +27,7 @@ network:
 - {tech: 2G, bands: '' }
 - {tech: 3G, bands: '' }
 - {tech: 4G, bands: '' }
-peripherals: ['']
+peripherals: None
 ram: 2 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> + <kbd>Power</kbd>.
   On the next screen use <kbd>Volume Down</kbd> to scroll to recovery and then press <kbd>Volume Up</kbd> to select.
diff --git a/_data/devices/m216.yml b/_data/devices/m216.yml
index 7b3f3c0..65c3a9e 100644
--- a/_data/devices/m216.yml
+++ b/_data/devices/m216.yml
@@ -26,7 +26,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '850 1900 2100 MHz UMTS' }
 - {tech: 4G, bands: 'B5(850) B3(1800 +) B40(TD 2300) LTE' }
-peripherals: [Accelerometer, Gyroscope, Proximity sensor, Magnetic sensor, Digital compass, GPS, NFC, FM Radio, RGB ambient light, Hall sensor]
+peripherals: [Accelerometer, Gyroscope, Proximity sensor, Hall effect sensor, Compass, GPS, NFC, FM radio, RGB ambient light]
 ram: 1.5 GB / 2 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> + <kbd>Power</kbd> until the LG logo appears, then release <kbd>Power</kbd> for a second and hold it again until the recovery comes up.
 recovery_partition: /dev/block/bootdevice/by-name/recovery
diff --git a/_data/devices/m7.yml b/_data/devices/m7.yml
index 52262bb..ae2a51f 100644
--- a/_data/devices/m7.yml
+++ b/_data/devices/m7.yml
@@ -25,7 +25,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '850 900 1900 2100 MHz HSDPA' }
 - {tech: 4G, bands: '800 1800 2600 MHz (International); 700 MHz and AWS Band (North America)' }
-peripherals: [Dual front speakers, Accelerometer, Gyroscope, Proximity sensor, Compass, GPS, Hall sensor, NFC, IR blaster]
+peripherals: [Dual front speakers, Accelerometer, Gyroscope, Proximity sensor, Compass, GPS, Hall effect sensor, NFC, IR blaster]
 ram: 2 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> + <kbd>Power</kbd>.
 release: 2013-03
diff --git a/_data/devices/m7vzw.yml b/_data/devices/m7vzw.yml
index f221303..f5048af 100644
--- a/_data/devices/m7vzw.yml
+++ b/_data/devices/m7vzw.yml
@@ -25,7 +25,7 @@ network:
 - {tech: 2G, bands: '800 1900 MHz CDMA' }
 - {tech: 3G, bands: 'CDMA2000 1xEV-DO' }
 - {tech: 4G, bands: '13(700)' }
-peripherals: [Dual front speakers, Accelerometer, Gyroscope, Proximity sensor, Compass, GPS, Hall sensor, NFC, IR blaster]
+peripherals: [Dual front speakers, Accelerometer, Gyroscope, Proximity sensor, Compass, GPS, Hall effect sensor, NFC, IR blaster]
 ram: 2 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> + <kbd>Power</kbd>.
 release: 2013-03
diff --git a/_data/devices/m8.yml b/_data/devices/m8.yml
index 658beda..f7d8bbb 100644
--- a/_data/devices/m8.yml
+++ b/_data/devices/m8.yml
@@ -26,7 +26,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '850 1900 2100 MHz HSDPA (Sprint), 850 1900 2100 MHz HSDPA (AT&T), 850 1700 1900 2100 MHz (T-Mobile), 850 900 1900 2100 MHz HSDPA (Verizon Wireless, EMEA, Asia), CDMA2000 1xEV-DO (Sprint, Verizon Wireless)' }
 - {tech: 4G, bands: '3(1800) 7(2600) 8(900) 20(800) LTE (EMEA), 1(2100) 3(1800) 7(2600) 8(900) 28(700) LTE (Asia), 4(1700/2100) 17(700) LTE (T-Mobile), 2(1900) 4(1700/2100) 5(850) 7(2600) 17(700) LTE (AT&T), 2(1900) 7(2600) 20(800) LTE (Sprint), 1(2100) 3(1800) 7(2600) 28(700) 4(1700) LTE (Verizon Wireless)' }
-peripherals: [Dual front speakers, Accelerometer, Gyroscope, Proximity sensor, Compass, GPS, Hall sensor, NFC, IR blaster]
+peripherals: [Dual front speakers, Accelerometer, Gyroscope, Proximity sensor, Compass, GPS, Hall effect sensor, NFC, IR blaster]
 ram: 2 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> + <kbd>Power</kbd> until the bootloader appears, then release the buttons.
   Use the volume buttons to scroll to "RECOVERY" and press <kbd>Power</kbd> to select.
diff --git a/_data/devices/m8d.yml b/_data/devices/m8d.yml
index e7acd86..20b8b1a 100644
--- a/_data/devices/m8d.yml
+++ b/_data/devices/m8d.yml
@@ -27,7 +27,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '850 900 1900 2100 MHz HSDPA (SIM 1 only)' }
 - {tech: 4G, bands: '3(1800) 7(2600) 8(900) 20(800) LTE' }
-peripherals: [Dual front speakers, Accelerometer, Gyroscope, Proximity sensor, Compass, GPS, Hall sensor, NFC, IR blaster]
+peripherals: [Dual front speakers, Accelerometer, Gyroscope, Proximity sensor, Compass, GPS, Hall effect sensor, NFC, IR blaster]
 ram: 2 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> + <kbd>Power</kbd> until the bootloader appears, then release the buttons.
   Use the volume buttons to scroll to "RECOVERY" and press <kbd>Power</kbd> to select.
diff --git a/_data/devices/manta.yml b/_data/devices/manta.yml
index e606f1a..315ad37 100644
--- a/_data/devices/manta.yml
+++ b/_data/devices/manta.yml
@@ -24,7 +24,7 @@ network:
 - {tech: 3G, bands: '' }
 - {tech: 4G, bands: '' }
 no_oem_unlock_switch: true
-peripherals: [Accelerometer, Gyroscope, Proximity sensor, Compass, GPS, Hall sensor, Barometer, Dual side NFC]
+peripherals: [Accelerometer, Gyroscope, Proximity sensor, Compass, GPS, Hall effect sensor, Barometer, Dual side NFC]
 ram: 2 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> + <kbd>Power</kbd>.
 release: 2012-11-13
diff --git a/_data/devices/maserati.yml b/_data/devices/maserati.yml
index 01898fa..9c886a3 100644
--- a/_data/devices/maserati.yml
+++ b/_data/devices/maserati.yml
@@ -24,7 +24,7 @@ network:
 - {tech: 2G, bands: '800 1900 MHz CDMA' }
 - {tech: 3G, bands: 'CDMA2000 1xEV-DO' }
 - {tech: 4G, bands: '13(700)' }
-peripherals: [Accelerometer, Proximity sensor, Compass, GPS, HDMI, FM Radio]
+peripherals: [Accelerometer, Proximity sensor, Compass, GPS, HDMI, FM radio]
 ram: 1 GB
 recovery_boot: With the device powered off, press <kbd>Power</kbd>. At the SafeStrap splash screen press Menu within 10 seconds.
 release: 2012-02-10
diff --git a/_data/devices/mido.yml b/_data/devices/mido.yml
index 8248594..5025e0d 100644
--- a/_data/devices/mido.yml
+++ b/_data/devices/mido.yml
@@ -26,7 +26,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '850 900 1900 2100 MHz HSDPA' }
 - {tech: 4G, bands: '1(2100) 3(1800) 4(1700/2100) 5(850) 7(2600) 8(900) 20(800) 38(2600) 40(2300), 3(1800) 5(850) 40(2300) (India)' }
-peripherals: [Fingerprint reader, Accelerometer, Gyroscope, Proximity sensor, Compass, GPS, Infrared sensor, FM Radio]
+peripherals: [Fingerprint reader, Accelerometer, Gyroscope, Proximity sensor, Compass, GPS, Infrared sensor, FM radio]
 ram: 2/3/4 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Up</kbd> + <kbd>Power</kbd>.
   Keep holding both buttons until the "MI" logo appears on the screen, then release.
diff --git a/_data/devices/mondrianwifi.yml b/_data/devices/mondrianwifi.yml
index d7f0ecd..5fb43e6 100644
--- a/_data/devices/mondrianwifi.yml
+++ b/_data/devices/mondrianwifi.yml
@@ -23,7 +23,7 @@ network:
 - {tech: 2G, bands: '' }
 - {tech: 3G, bands: '' }
 - {tech: 4G, bands: '' }
-peripherals: ['']
+peripherals: None
 ram: 2 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Up</kbd> while booting.
 release: 2014-01
diff --git a/_data/devices/n3.yml b/_data/devices/n3.yml
index 531744e..3da1251 100644
--- a/_data/devices/n3.yml
+++ b/_data/devices/n3.yml
@@ -25,7 +25,7 @@ network:
 - {tech: 3G, bands: '' }
 - {tech: 4G, bands: '' }
 no_oem_unlock_switch: true
-peripherals: ['']
+peripherals: None
 ram: 2 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> + <kbd>Power</kbd>.
 release: 2015-01
diff --git a/_data/devices/nx512j.yml b/_data/devices/nx512j.yml
index 90f21e3..255799b 100644
--- a/_data/devices/nx512j.yml
+++ b/_data/devices/nx512j.yml
@@ -25,7 +25,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '850 900 1900 2100 MHz HSDPA' }
 - {tech: 4G, bands: 'B1(2100) B3(1800) B3(2600) B38(TDD 2600) B39(TDD 19000) B40(TDD 2300) B41(TDD 2500)MHz' }
-peripherals: [Dual SIM, FM Radio, Accelerometer, Gyroscope, Proximity sensor, Compass, Hall sensor]
+peripherals: [Dual SIM, FM radio, Accelerometer, Gyroscope, Proximity sensor, Compass, Hall effect sensor]
 ram: 2 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Up</kbd> + <kbd>Power</kbd>.
 release: 2015-06
diff --git a/_data/devices/onyx.yml b/_data/devices/onyx.yml
index f30296c..72d4ea3 100644
--- a/_data/devices/onyx.yml
+++ b/_data/devices/onyx.yml
@@ -23,7 +23,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '850 900 1900 2100 MHz HSDPA (Europe), 850 900 1700(AWS) 1900 2100 MHz (USA)' }
 - {tech: 4G, bands: '1(2100) 3(1800) 5(850) 7(2600) 8(900) 20(800) 38(2600) 40(2300) LTE (Europe), 1(2100) 2(1900) 4(1700/2100) 5(850) 7(2600) 8(900) LTE (USA)' }
-peripherals: [Accelerometer, Gyroscope, Proximity sensor, Dual SIM, GPS, FM Radio]
+peripherals: [Accelerometer, Gyroscope, Proximity sensor, Dual SIM, GPS, FM radio]
 ram: 3 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> + <kbd>Power</kbd>.
 release: 2015-11
diff --git a/_data/devices/paella.yml b/_data/devices/paella.yml
index c302a54..c28515a 100644
--- a/_data/devices/paella.yml
+++ b/_data/devices/paella.yml
@@ -26,7 +26,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '900 2100 MHz HSDPA' }
 - {tech: 4G, bands: '800 1800 2100 2600 MHz LTE' }
-peripherals: [Dual SIM, Accelerometer, Gyroscope, Light sensor, Proximity sensor, Compass, Hall sensor, GPS, FM Radio, USB OTG]
+peripherals: [Dual SIM, Accelerometer, Gyroscope, Light sensor, Proximity sensor, Compass, Hall effect sensor, GPS, FM radio, USB OTG]
 ram: 2/3 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Up</kbd> + <kbd>Power</kbd>.
 release: 2015
diff --git a/_data/devices/piccolo.yml b/_data/devices/piccolo.yml
index b0f5b1a..b4fdf2d 100644
--- a/_data/devices/piccolo.yml
+++ b/_data/devices/piccolo.yml
@@ -26,7 +26,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '900 2100 MHz HSDPA' }
 - {tech: 4G, bands: '800 1800 2100 2600 MHz LTE' }
-peripherals: [Dual SIM, Accelerometer, Gyroscope, Light sensor, Proximity sensor, Compass, Hall sensor, GPS, FM Radio, NFC, USB OTG]
+peripherals: [Dual SIM, Accelerometer, Gyroscope, Light sensor, Proximity sensor, Compass, Hall effect sensor, GPS, FM radio, NFC, USB OTG]
 ram: 2/3 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Up</kbd> + <kbd>Power</kbd>.
   When the device vibrates, release <kbd>Power</kbd>, wait a second, then release <kbd>Volume Up</kbd>.
diff --git a/_data/devices/pme.yml b/_data/devices/pme.yml
index db3615a..6efaa44 100644
--- a/_data/devices/pme.yml
+++ b/_data/devices/pme.yml
@@ -25,7 +25,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM, 800 1900 MHz CDMA (USA)' }
 - {tech: 3G, bands: '850 900 1900 2100 MHz HSDPA (EMEA), 850 900 1700(AWS) 1900 2100 MHz HSDPA (USA)' }
 - {tech: 4G, bands: '1(2100) 3(1800) 5(850) 7(2600) 8(900) 20(800) 28(700) 38(2600) 40(2300) 41(2500) (EMEA), 1(2100) 2(1900) 3(1800) 4(1700/2100) 5(850) 7(2600) 12(700) 13(700) 17(700) 20(800) 28(700) 29(700) 30(2300) (USA)' }
-peripherals: [Iceview Cases, Hall sensor]
+peripherals: [Iceview Cases, Hall effect sensor]
 ram: 4 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> while booting.
   Select "Reboot to bootloader" and then in the bootloader select "Boot to recovery mode".
diff --git a/_data/devices/r5.yml b/_data/devices/r5.yml
index 3891805..094face 100644
--- a/_data/devices/r5.yml
+++ b/_data/devices/r5.yml
@@ -28,7 +28,7 @@ network:
 - {tech: 3G, bands: '850 900 1700 1900 2100 MHz HSPA+ (R5), 850 900 1900 2100 MHz HSPA+ (R5s)' }
 - {tech: 4G, bands: '1(2100) 3(1800) 4(1700/2100) 7(2600) 8(900) 17(700) 20(800) 28(700) 40(2300) LTE(Cat. 4) (R5), 1(2100) 3(1800) 7(2600) 40(2300) LTE(Cat. 4) (R5s)' }
 no_oem_unlock_switch: true
-peripherals: [A-GPS, Accelerometer, Compass, GLONASS, GPS, Gyroscope, Light sensor, Magnetic sensor, Proximity sensor, USB OTG]
+peripherals: [A-GPS, Accelerometer, Compass, GLONASS, GPS, Gyroscope, Light sensor, Proximity sensor, USB OTG]
 project_spectrum: "https://androidfilehost.com/?fid=962021903579484169"
 project_spectrum_recovery: "https://androidfilehost.com/?fid=889964283620757488"
 ram: 2 GB (R5), 3 GB (R5s)
diff --git a/_data/devices/r7plus.yml b/_data/devices/r7plus.yml
index e131826..10c3b85 100644
--- a/_data/devices/r7plus.yml
+++ b/_data/devices/r7plus.yml
@@ -27,7 +27,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '850 900 1900 2100 MHz HSPA+' }
 - {tech: 4G, bands: '1(2100) 3(1800) 5(850) 7(2600) 8(900) 40(2300) LTE(Cat. 4)' }
-peripherals: [A-GPS, Accelerometer, Compass, Dual SIM, GLONASS, GPS, Gyroscope, Light sensor, Magnetic sensor, Proximity sensor, USB OTG, Fingerprint reader]
+peripherals: [A-GPS, Accelerometer, Compass, Dual SIM, GLONASS, GPS, Gyroscope, Light sensor, Proximity sensor, USB OTG, Fingerprint reader]
 ram: 3/4 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> + <kbd>Power</kbd>.
 release: 2015-05
diff --git a/_data/devices/r7sf.yml b/_data/devices/r7sf.yml
index d8a29c1..8b987fd 100644
--- a/_data/devices/r7sf.yml
+++ b/_data/devices/r7sf.yml
@@ -27,7 +27,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '850 900 1900 2100 MHz HSPA+' }
 - {tech: 4G, bands: '1(2100) 3(1800) 5(850) 7(2600) 8(900) 20(800) 40(2300) LTE(Cat. 4)' }
-peripherals: [A-GPS, Accelerometer, Compass, Dual SIM, GLONASS, GPS, Gyroscope, Light sensor, Magnetic sensor, Proximity sensor, USB OTG]
+peripherals: [A-GPS, Accelerometer, Compass, Dual SIM, GLONASS, GPS, Gyroscope, Light sensor, Proximity sensor, USB OTG]
 project_spectrum: "https://androidfilehost.com/?fid=673791459329053792"
 ram: 4 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> + <kbd>Power</kbd>.
diff --git a/_data/devices/roth.yml b/_data/devices/roth.yml
index 2d092ce..6059594 100644
--- a/_data/devices/roth.yml
+++ b/_data/devices/roth.yml
@@ -25,7 +25,7 @@ network:
 - {tech: 3G, bands: '' }
 - {tech: 4G, bands: '' }
 no_oem_unlock_switch: true
-peripherals: ['']
+peripherals: None
 ram: 2 GB
 recovery_boot: With the device powered off, hold Back + <kbd>Home</kbd> and tap <kbd>Power</kbd>,
   then release once menu is displayed. Tap <kbd>Home</kbd> until recovery is selected. Press <kbd>Power</kbd> to select.
diff --git a/_data/devices/serrano3gxx.yml b/_data/devices/serrano3gxx.yml
index 204efe6..fe689cd 100644
--- a/_data/devices/serrano3gxx.yml
+++ b/_data/devices/serrano3gxx.yml
@@ -27,7 +27,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '850 900 1900 2100 MHz HSDPA' }
 - {tech: 4G, bands: '' }
-peripherals: [Accelerometer, Gyroscope, Proximity sensor, Compass, SNS integration, FM Radio, Infrared sensor, GPS, GLONASS]
+peripherals: [Accelerometer, Gyroscope, Proximity sensor, Compass, SNS integration, FM radio, Infrared sensor, GPS, GLONASS]
 ram: 1.5 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Up</kbd> + <kbd>Home</kbd> + <kbd>Power</kbd>.
 release: 2013-07
diff --git a/_data/devices/serranodsdd.yml b/_data/devices/serranodsdd.yml
index 8292a3d..89c982d 100644
--- a/_data/devices/serranodsdd.yml
+++ b/_data/devices/serranodsdd.yml
@@ -27,7 +27,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '850 900 1900 2100 MHz HSDPA' }
 - {tech: 4G, bands: '' }
-peripherals: [Dual SIM, Accelerometer, Gyroscope, Proximity sensor, Compass, SNS integration, FM Radio, Infrared sensor, GPS, GLONASS]
+peripherals: [Dual SIM, Accelerometer, Gyroscope, Proximity sensor, Compass, SNS integration, FM radio, Infrared sensor, GPS, GLONASS]
 ram: 1.5 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Up</kbd> + <kbd>Home</kbd> + <kbd>Power</kbd>.
 release: 2013-07
diff --git a/_data/devices/serranoltexx.yml b/_data/devices/serranoltexx.yml
index 3a2431c..bcaf9d1 100644
--- a/_data/devices/serranoltexx.yml
+++ b/_data/devices/serranoltexx.yml
@@ -27,7 +27,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '850 900 1900 2100 MHz HSDPA' }
 - {tech: 4G, bands: '3(1800) 7(2600) 8(900) 20(800)' }
-peripherals: [Accelerometer, Gyroscope, Proximity sensor, Compass, SNS integration, FM Radio, Infrared sensor, GPS, GLONASS, NFC]
+peripherals: [Accelerometer, Gyroscope, Proximity sensor, Compass, SNS integration, FM radio, Infrared sensor, GPS, GLONASS, NFC]
 ram: 1.5 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Up</kbd> + <kbd>Home</kbd> + <kbd>Power</kbd>.
 release: 2013-07
diff --git a/_data/devices/shamu.yml b/_data/devices/shamu.yml
index f1d2544..ca9228d 100644
--- a/_data/devices/shamu.yml
+++ b/_data/devices/shamu.yml
@@ -25,7 +25,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM (All models), 800 1900 MHz CDMA (North/South American model only)' }
 - {tech: 3G, bands: '850 900 1700 1900 2100 MHz HSDPA (North/South American model only), 800 850 900 1700 1800 1900 2100 MHz HSDPA (Afro-Eurasian models)' }
 - {tech: 4G, bands: '2(1900) 3(1800) 4(1700/2100) 5(850) 7(2600) 12(700) 13(700) 17(700) 25(1900) 26(850) 29(700) 41(2500) (North/South American models), 1(2100) 3(1800) 5(850) 7(2600) 8(900) 9(1800) 19(800) 20(800) 28(700) 41(2500) MHz (Afro-Eurasian models)' }
-peripherals: [Accelerometer, Gyroscope, Proximity sensor, Compass, Barometer, Hall sensor, NFC, Qi wireless charging, Dual front speakers]
+peripherals: [Accelerometer, Gyroscope, Proximity sensor, Compass, Barometer, Hall effect sensor, NFC, Qi wireless charging, Dual front speakers]
 ram: 3 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> while booting. Then use the menu to select Recovery.
 release: 2014-10-29
diff --git a/_data/devices/shieldtablet.yml b/_data/devices/shieldtablet.yml
index 10445f8..df98d8a 100644
--- a/_data/devices/shieldtablet.yml
+++ b/_data/devices/shieldtablet.yml
@@ -27,7 +27,7 @@ network:
 - {tech: 3G, bands: '850 1700 1900 2100 MHz HSDPA (North America), 850 900 1900 2100 HSDPA (International, depending on model)' }
 - {tech: 4G, bands: '2(1900) 4(1700/2100) 5(850) 7(2600) 17(700) (North America), 1(2100) 3(1800) 7(2600) 20(800) (International, depending on model)' }
 no_oem_unlock_switch: true
-peripherals: ['']
+peripherals: None
 ram: 2 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> and tap <kbd>Power</kbd>,
   then release once menu is displayed. Tap <kbd>Volume Down</kbd> until recovery is selected. Press <kbd>Power</kbd> to select.
diff --git a/_data/devices/sumire.yml b/_data/devices/sumire.yml
index f7a665c..5853d3e 100644
--- a/_data/devices/sumire.yml
+++ b/_data/devices/sumire.yml
@@ -27,8 +27,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '850 900 1700(AWS) 1900 2100 MHz HSDPA' }
 - {tech: 4G, bands: '1(2100) 2(1900) 3(1800) 4(1700/2100) 5(850) 7(2600) 8(900) 12(700) 17(700) 20(800) LTE' }
-peripherals: [Wi-Fi Direct, Wi-Fi Display, Wi-Fi Hotspot, MHL HDMI, GPS, NFC, Accelerometer,
-  Compass, Gyroscope, Light sensor, Proximity Sensor, Fingerprint]
+peripherals: [Wi-Fi Display, MHL HDMI, GPS, NFC, Accelerometer, Compass, Gyroscope, Light sensor, Proximity sensor, Fingerprint reader]
 ram: 3 GB
 recovery_boot: With the device powered down, hold the <kbd>Volume Down</kbd> and <kbd>Power</kbd> buttons.
 release: 2015-09
diff --git a/_data/devices/suzuran.yml b/_data/devices/suzuran.yml
index 3261953..a8cefa2 100644
--- a/_data/devices/suzuran.yml
+++ b/_data/devices/suzuran.yml
@@ -27,7 +27,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '850 900 1900 2100 MHz HSDPA' }
 - {tech: 4G, bands: '1 (2100) 2 (1900) 3 (1800) 4 (1700/2100) 5 (850) 7 (2600) 8 (900) 12 (700) 17 (700) 20 (800) 28 (700) 38 (2600) 40 (2300) LTE' }
-peripherals: [Accelerometer, Fingerprint reader, Gyroscope, Proximity sensor, Ambient light, GPS, NFC]
+peripherals: [Accelerometer, Fingerprint reader, Gyroscope, Proximity sensor, Light sensor, GPS, NFC]
 ram: 2 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> + <kbd>Power</kbd>.
 release: 2015-10
diff --git a/_data/devices/t6.yml b/_data/devices/t6.yml
index c687833..5ac5e7a 100644
--- a/_data/devices/t6.yml
+++ b/_data/devices/t6.yml
@@ -26,7 +26,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '900 1900 2100 MHz HSDPA (EMEA), 850 900 1900 2100 MHz HSDPA (Asia)' }
 - {tech: 4G, bands: '3(1800) 7(2600) 8(900) 20(800) (EMEA), 1(2100) 3(1800) 7(2600) 8(900) (Asia)' }
-peripherals: [Dual front speakers, Accelerometer, Gyroscope, Proximity sensor, Compass, GPS, Hall sensor, NFC, IR blaster]
+peripherals: [Dual front speakers, Accelerometer, Gyroscope, Proximity sensor, Compass, GPS, Hall effect sensor, NFC, IR blaster]
 ram: 2 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> + <kbd>Power</kbd> until the bootloader appears, then release the buttons.
   Use the volume buttons to scroll to "RECOVERY" and press <kbd>Power</kbd> to select.
diff --git a/_data/devices/t6vzw.yml b/_data/devices/t6vzw.yml
index dfcf541..99abed6 100644
--- a/_data/devices/t6vzw.yml
+++ b/_data/devices/t6vzw.yml
@@ -26,7 +26,7 @@ network:
 - {tech: 2G, bands: '800 1900 MHz CDMA' }
 - {tech: 3G, bands: '1900 2100MHz HSPA' }
 - {tech: 4G, bands: '4(1700/2100) 13(700) LTE' }
-peripherals: [Dual front speakers, Accelerometer, Gyroscope, Proximity sensor, Compass, GPS, Hall sensor, NFC, IR blaster]
+peripherals: [Dual front speakers, Accelerometer, Gyroscope, Proximity sensor, Compass, GPS, Hall effect sensor, NFC, IR blaster]
 ram: 2 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> + <kbd>Power</kbd> until the bootloader appears, then release the buttons.
   Use the volume buttons to scroll to "RECOVERY" and press <kbd>Power</kbd> to select.
diff --git a/_data/devices/tenshi.yml b/_data/devices/tenshi.yml
index 7285ff9..c57c56a 100644
--- a/_data/devices/tenshi.yml
+++ b/_data/devices/tenshi.yml
@@ -25,7 +25,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '900 2100 MHz HSDPA' }
 - {tech: 4G, bands: '800 1800 2100 2600 MHz LTE' }
-peripherals: [Dual SIM, Accelerometer, Gyroscope, Light sensor, Proximity sensor, Compass, Hall sensor, GPS, FM Radio, USB OTG]
+peripherals: [Dual SIM, Accelerometer, Gyroscope, Light sensor, Proximity sensor, Compass, Hall effect sensor, GPS, FM radio, USB OTG]
 ram: 2/3 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Up</kbd> + <kbd>Power</kbd>.
 release: 2016
diff --git a/_data/devices/us996.yml b/_data/devices/us996.yml
index fd38d5f..5092a39 100644
--- a/_data/devices/us996.yml
+++ b/_data/devices/us996.yml
@@ -26,7 +26,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '850 900 1900 2100 MHz HSDPA' }
 - {tech: 4G, bands: '1(2100) 17(700) 5(850) 4(1700/2100) 3(1800) 7(2600) 2(1900) 8(900) 20(800) 28(700)' }
-peripherals: [Fingerprint reader, Accelerometer, Gyroscope, Proximity sensor, Compass, Barometer, Color spectrum, Ambient light sensor, GPS, IR blaster, Infrared sensor, Mobility DisplayPort (MyDP), NFC, USB OTG]
+peripherals: [Fingerprint reader, Accelerometer, Gyroscope, Proximity sensor, Compass, Barometer, Color spectrum sensor, Light sensor, GPS, IR blaster, Infrared sensor, Mobility DisplayPort (MyDP), NFC, USB OTG]
 ram: 4 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> + <kbd>Power</kbd> until the LG logo appears, then release <kbd>Power</kbd> for a second and hold it again until the recovery comes up.
 release: 2016-10
diff --git a/_data/devices/us997.yml b/_data/devices/us997.yml
index b74cf4f..56eb152 100644
--- a/_data/devices/us997.yml
+++ b/_data/devices/us997.yml
@@ -26,7 +26,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '850 900 1900 2100 MHz HSDPA' }
 - {tech: 4G, bands: '1(2100) 17(700) 5(850) 4(1700/2100) 3(1800) 7(2600) 2(1900) 8(900) 20(800) 28(700)' }
-peripherals: [Fingerprint reader, Accelerometer, Barometer, Gyroscope, Proximity sensor, Compass, Barometer, Color spectrum]
+peripherals: [Fingerprint reader, Accelerometer, Barometer, Gyroscope, Proximity sensor, Compass, Color spectrum sensor]
 ram: 4 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> + <kbd>Power</kbd> until the LG logo appears, then release <kbd>Power</kbd> for a second and hold it again until the recovery comes up.
 release: 2017-02
diff --git a/_data/devices/v400.yml b/_data/devices/v400.yml
index e8a4b92..ea951b5 100644
--- a/_data/devices/v400.yml
+++ b/_data/devices/v400.yml
@@ -22,7 +22,7 @@ install_method: dd
 kernel: android_kernel_lge_v4xx
 maintainers: [mycax]
 name: G Pad 7.0 WiFi
-peripherals: ['']
+peripherals: None
 ram: 1 GB
 recovery_boot: With the device powered off, hold <kbd>Power</kbd> + <kbd>Volume Down</kbd>. When the LG logo appears, release both buttons and then hold them again.
   When prompted to erase all user data, choose 'Yes'. Once again choose 'Yes' to confirm and then device will boot to recovery.
diff --git a/_data/devices/v410.yml b/_data/devices/v410.yml
index 870b9cb..a25960e 100644
--- a/_data/devices/v410.yml
+++ b/_data/devices/v410.yml
@@ -26,7 +26,7 @@ network:
 - {tech: 2G, bands: '' }
 - {tech: 3G, bands: '' }
 - {tech: 4G, bands: '' }
-peripherals: ['']
+peripherals: None
 ram: 1 GB
 recovery_boot: With the device powered off, hold <kbd>Power</kbd> + <kbd>Volume Down</kbd>. When the LG logo appears, release both buttons and then hold them again.
   When prompted to erase all user data, choose 'Yes'. Once again choose 'Yes' to confirm and then device will boot to recovery.
diff --git a/_data/devices/vegetalte.yml b/_data/devices/vegetalte.yml
index 1de0bb9..6c0ddfc 100644
--- a/_data/devices/vegetalte.yml
+++ b/_data/devices/vegetalte.yml
@@ -29,7 +29,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '900 2100 MHz HSDPA' }
 - {tech: 4G, bands: '800 1800 2600 MHz LTE' }
-peripherals: [Dual SIM, Accelerometer, Gyroscope, Light sensor, Proximity sensor, Compass, GPS, FM Radio, USB OTG]
+peripherals: [Dual SIM, Accelerometer, Gyroscope, Light sensor, Proximity sensor, Compass, GPS, FM radio, USB OTG]
 ram: 1/2 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Up</kbd> + <kbd>Power</kbd>.
   When the device vibrates, release <kbd>Power</kbd>, wait a second, then release <kbd>Volume Up</kbd>.
diff --git a/_data/devices/vs995.yml b/_data/devices/vs995.yml
index c65c836..581aa86 100644
--- a/_data/devices/vs995.yml
+++ b/_data/devices/vs995.yml
@@ -26,7 +26,7 @@ network:
 - {tech: 2G, bands: '850 900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '850 900 1900 2100 MHz HSDPA' }
 - {tech: 4G, bands: '1(2100) 17(700) 5(850) 4(1700/2100) 3(1800) 7(2600) 2(1900) 8(900) 20(800) 28(700)' }
-peripherals: [Fingerprint reader, Accelerometer, Gyroscope, Proximity sensor, Compass, Barometer, Color spectrum, Ambient light sensor, GPS, IR blaster, Infrared sensor, Mobility DisplayPort (MyDP), NFC, USB OTG]
+peripherals: [Fingerprint reader, Accelerometer, Gyroscope, Proximity sensor, Compass, Barometer, Color spectrum sensor, Light sensor, GPS, IR blaster, Infrared sensor, Mobility DisplayPort (MyDP), NFC, USB OTG]
 ram: 4 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> + <kbd>Power</kbd> until the LG logo appears, then release <kbd>Power</kbd> for a second and hold it again until the recovery comes up.
 release: 2016-10
diff --git a/_data/devices/w5.yml b/_data/devices/w5.yml
index 4b409f4..67a9a3d 100644
--- a/_data/devices/w5.yml
+++ b/_data/devices/w5.yml
@@ -22,7 +22,7 @@ network:
 - {tech: 2G, bands: '900 1800 1900 MHz GSM' }
 - {tech: 3G, bands: '900 2100 MHz HSDPA' }
 - {tech: 4G, bands: '' }
-peripherals: [GPS, Accelerometer, Compass, Proximity sensor, FM Radio]
+peripherals: [GPS, Accelerometer, Compass, Proximity sensor, FM radio]
 ram: 1 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Down</kbd> + <kbd>Power</kbd>.
   Release the <kbd>Power</kbd> button for a brief second when the LG logo is displayed, then
diff --git a/_data/devices/zerofltexx.yml b/_data/devices/zerofltexx.yml
index 1851117..f81443b 100644
--- a/_data/devices/zerofltexx.yml
+++ b/_data/devices/zerofltexx.yml
@@ -28,7 +28,7 @@ network:
 - {tech: 4G, bands: 'B1-B8, B12, B17-B20, B26' }
 peripherals: [Fingerprint reader, Heart rate sensor, Motion Coprocessor, Accelerometer,
   Gesture sensor, Gyroscope, Proximity sensor, Compass, Barometer, Hall effect sensor,
-  Magnetic sensor, RGB ambient light, Infrared (IR) LED sensor, GPS, NFC]
+  RGB ambient light, Infrared sensor, GPS, NFC]
 ram: 3 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Up</kbd> + <kbd>Home</kbd> + <kbd>Power</kbd>.
   When the blue text appears, release the buttons.
diff --git a/_includes/deviceinfo.html b/_includes/deviceinfo.html
index 1947f3b..704f776 100644
--- a/_includes/deviceinfo.html
+++ b/_includes/deviceinfo.html
@@ -147,15 +147,23 @@ ul, ol {
         <td>{{ device.wifi }}</td>
     </tr>
     {% endif %}
-    {% if device.peripherals[0] != '' %}
+    {% if device.peripherals != 'None' %}
     <tr>
         <td class="header">Peripherals</td>
         <td>
-            <ul>
-                {% for el in device.peripherals %}
-                <li>{{ el }}</li>
-                {% endfor %}
-            </ul>
+            {%- if device.peripherals.first.size == 1 %}
+                {%- for model in device.peripherals %}
+                <ul>
+                    <li>{{ model.first[0] }}
+                        {% assign peripherals = model.first[1] %}
+                        {% include snippets/peripherals.html %}
+                    </li>
+                </ul>
+                {%- endfor %}
+            {%- else %}
+                {%- assign peripherals = device.peripherals %}
+                {%- include snippets/peripherals.html %}
+            {%- endif %}
         </td>
     </tr>
     {% endif %}
diff --git a/_includes/snippets/peripherals.html b/_includes/snippets/peripherals.html
new file mode 100644
index 0000000..959b8d5
--- /dev/null
+++ b/_includes/snippets/peripherals.html
@@ -0,0 +1,5 @@
+<ul>
+    {%- for peripheral in peripherals %}
+    <li>{{ peripheral }}</li>
+    {%- endfor %}
+</ul>
\ No newline at end of file
diff --git a/device_sample/sample.yml b/device_sample/sample.yml
index a27b557..f362627 100644
--- a/device_sample/sample.yml
+++ b/device_sample/sample.yml
@@ -31,7 +31,7 @@ network:
 - {tech: 4G, bands: '3(1800) 7(2600) 8(900) 20(800) LTE' }
 peripherals: [Fingerprint reader, Heart rate sensor, Motion Coprocessor, Accelerometer,
   Gesture sensor, Gyroscope, Proximity sensor, Compass, Barometer, Hall effect sensor,
-  Magnetic sensor, RGB ambient light, Infrared (IR) LED sensor, GPS, NFC]
+  RGB ambient light, Infrared sensor, GPS, NFC]
 ram: 2 GB
 recovery_boot: With the device powered off, hold <kbd>Volume Up</kbd> + <kbd>Home</kbd> + <kbd>Power</kbd>.
   When the blue text appears, release the buttons.
diff --git a/test/schema-06.yml b/test/schema-06.yml
index d631a56..e53cc83 100644
--- a/test/schema-06.yml
+++ b/test/schema-06.yml
@@ -58,6 +58,52 @@ definitions:
     - capacity
     - removable
     type: object
+  valid_peripherals:
+    description: All valid peripherals a device can have
+    items:
+      enum:
+      - A-GPS
+      - Accelerometer
+      - Audio Coprocessor
+      - Barometer
+      - Color spectrum sensor
+      - Compass
+      - Dual SIM
+      - Dual front speakers
+      - Dual side NFC
+      - FM radio
+      - Fingerprint reader
+      - GLONASS
+      - GPS
+      - Gesture sensor
+      - Gyroscope
+      - HDMI
+      - Hall effect sensor
+      - Hardware keyboard
+      - Heart rate sensor
+      - Hygrometer
+      - IR blaster
+      - Iceview Cases
+      - Infrared sensor
+      - Light sensor
+      - MHL
+      - MHL 2.0
+      - MHL HDMI
+      - Mobility DisplayPort (MyDP)
+      - Motion Coprocessor
+      - NFC
+      - Pedometer
+      - Proximity sensor
+      - Qi wireless charging
+      - RGB ambient light
+      - SNS integration
+      - Sensor hub
+      - Thermometer
+      - USB OTG
+      - Wi-Fi Display
+      type: string
+    type: array
+    uniqueItems: True
 properties:
   architecture:
     "$id": "/properties/architecture"
@@ -238,6 +284,18 @@ properties:
     "$id": "/properties/name"
     minLength: 1
     type: string
+  peripherals:
+    "$id": /properties/peripherals
+    oneOf:
+    - "$ref": "#/definitions/valid_peripherals"
+    - items:
+        additionalProperties:
+          "$ref": "#/definitions/valid_peripherals"
+        description: Map of model number to valid peripherals
+        type: object
+      type: array
+    - enum:
+      - None
   release:
     "$id": "/properties/release"
     description: Date device released to public in ISO8601 reduced precision format
@@ -344,6 +402,7 @@ required:
 - kernel
 - maintainers
 - name
+- peripherals
 - release
 - type
 - vendor
-- 
2.17.1

